// moh-common-lib/esm2015/lib/components/postal-code/postal-code.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
import { MaskModel, LETTER, NUMBER, SPACE } from '../../../models/src/mask.model';
export class PostalCodeComponent extends MaskModel {
    constructor() {
        super();
        this.label = 'Postal Code';
        this.displayMask = true;
        this.maxlen = '250';
        this.placeholder = 'A1A 1A1';
        this.pcFormat = /^[A-Za-z][0-9][A-Za-z]\s?[0-9][A-Za-z][0-9]$/;
        this.mask = [LETTER, NUMBER, LETTER, SPACE, NUMBER, LETTER, NUMBER];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
PostalCodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'common-postal-code',
                template: "<label class=\"control-label\">{{label}}</label>\n<div *ngIf=\"displayMask; else NoMask\">\n<input #postalRef=\"ngModel\"\n       class=\"form-control\"\n       spellcheck=\"false\"\n       type=\"text\"\n       id=\"postalCode_{{objectId}}\"\n       name=\"postalCode_{{objectId}}\"\n       [placeholder]=\"placeholder\"\n       [ngModel]=\"value\"\n       (ngModelChange)=\"onUpdate($event)\"\n       [textMask]=\"{mask: mask, pipe: upperCasePipe}\"\n       [pattern]='pcFormat'\n       [required]=\"required\"\n       [disabled]=\"disabled\" />\n\n  <!-- Error messages for input -->\n  <div *ngIf=\"!disabled && postalRef.touched\"\n        role=\"alert\"\n        class=\"error-container\"\n        aria-live=\"assertive\">\n    <div class=\"text-danger\" *ngIf=\"postalRef?.errors?.required\">\n      {{label}} is required.\n    </div>\n    <div class=\"text-danger\" *ngIf=\"postalRef?.errors?.pattern\">\n      Must be in the format {{placeholder}}.\n    </div>\n  </div>\n</div>\n\n<ng-template #NoMask>\n  <input class=\"form-control\"\n         #postalRef=\"ngModel\"\n         spellcheck=\"false\"\n         type=\"text\"\n         id=\"postalCode_{{objectId}}\"\n         name=\"postalCode_{{objectId}}\"\n         [ngModel]=\"value\"\n         (ngModelChange)=\"onUpdate($event)\"\n         [required]=\"required\"\n         [disabled]=\"disabled\"\n         maxlength=\"{{maxlen}}\"/>\n\n  <!-- Error messages for input -->\n  <div *ngIf=\"!disabled && postalRef.touched\"\n        role=\"alert\"\n        class=\"error-container\"\n        aria-live=\"assertive\">\n    <div class=\"text-danger\" *ngIf=\"postalRef?.errors?.required\">\n      Postal Code is required.\n    </div>\n  </div>\n</ng-template>\n",
                // Re-use the same ngForm that it's parent is using. The component will show
                // up in its parents `this.form`, and will auto-update `this.form.valid`
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }],
                styles: [""]
            }] }
];
/** @nocollapse */
PostalCodeComponent.ctorParameters = () => [];
PostalCodeComponent.propDecorators = {
    label: [{ type: Input }],
    displayMask: [{ type: Input }],
    maxlen: [{ type: Input }],
    placeholder: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PostalCodeComponent.prototype.label;
    /** @type {?} */
    PostalCodeComponent.prototype.displayMask;
    /** @type {?} */
    PostalCodeComponent.prototype.maxlen;
    /** @type {?} */
    PostalCodeComponent.prototype.placeholder;
    /** @type {?} */
    PostalCodeComponent.prototype.pcFormat;
}
