// moh-common-lib/esm5/captcha/captcha-data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
// payload returned from the server
var ServerPayload = /** @class */ (function () {
    function ServerPayload() {
    }
    ServerPayload.decorators = [
        { type: Injectable }
    ];
    return ServerPayload;
}());
export { ServerPayload };
if (false) {
    /** @type {?} */
    ServerPayload.prototype.nonce;
    /** @type {?} */
    ServerPayload.prototype.captcha;
    /** @type {?} */
    ServerPayload.prototype.validation;
    /** @type {?} */
    ServerPayload.prototype.expiry;
}
var CaptchaDataService = /** @class */ (function () {
    function CaptchaDataService(httpClient) {
        this.httpClient = httpClient;
    }
    // private http: Http) { }
    // private http: Http) { }
    /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @return {?}
     */
    CaptchaDataService.prototype.fetchData = 
    // private http: Http) { }
    /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @return {?}
     */
    function (apiBaseUrl, nonce) {
        return this.httpClient
            .post(apiBaseUrl + '/captcha', { nonce: nonce }, { observe: 'response' });
    };
    /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @param {?} answer
     * @param {?} encryptedAnswer
     * @return {?}
     */
    CaptchaDataService.prototype.verifyCaptcha = /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @param {?} answer
     * @param {?} encryptedAnswer
     * @return {?}
     */
    function (apiBaseUrl, nonce, answer, encryptedAnswer) {
        return this.httpClient
            .post(apiBaseUrl + '/verify/captcha', { nonce: nonce, answer: answer, validation: encryptedAnswer }, { observe: 'response' });
    };
    /**
     * @param {?} apiBaseUrl
     * @param {?} validation
     * @param {?=} translation
     * @return {?}
     */
    CaptchaDataService.prototype.fetchAudio = /**
     * @param {?} apiBaseUrl
     * @param {?} validation
     * @param {?=} translation
     * @return {?}
     */
    function (apiBaseUrl, validation, translation) {
        /** @type {?} */
        var payload = { validation: validation };
        if (translation) {
            payload.translation = translation;
        }
        return this.httpClient
            .post(apiBaseUrl + '/captcha/audio', payload, { observe: 'response' });
    };
    CaptchaDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CaptchaDataService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return CaptchaDataService;
}());
export { CaptchaDataService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CaptchaDataService.prototype.httpClient;
}
