// moh-common-lib/esm5/captcha/captcha.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';
import { CaptchaComponent } from './captcha.component';
import { CaptchaDataService } from './captcha-data.service';
import { CommonModule } from '@angular/common';
var CaptchaModule = /** @class */ (function () {
    function CaptchaModule() {
    }
    /**
     * @return {?}
     */
    CaptchaModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: CaptchaModule,
            providers: [CaptchaDataService]
        };
    };
    CaptchaModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CaptchaComponent
                    ],
                    imports: [
                        HttpClientModule,
                        CommonModule,
                        FormsModule
                    ],
                    providers: [
                        CaptchaDataService
                    ],
                    exports: [
                        CaptchaComponent
                    ],
                },] }
    ];
    return CaptchaModule;
}());
export { CaptchaModule };
