// moh-common-lib/esm2015/captcha/captcha-data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
// payload returned from the server
export class ServerPayload {
}
ServerPayload.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    ServerPayload.prototype.nonce;
    /** @type {?} */
    ServerPayload.prototype.captcha;
    /** @type {?} */
    ServerPayload.prototype.validation;
    /** @type {?} */
    ServerPayload.prototype.expiry;
}
export class CaptchaDataService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    // private http: Http) { }
    /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @return {?}
     */
    fetchData(apiBaseUrl, nonce) {
        return this.httpClient
            .post(apiBaseUrl + '/captcha', { nonce: nonce }, { observe: 'response' });
    }
    /**
     * @param {?} apiBaseUrl
     * @param {?} nonce
     * @param {?} answer
     * @param {?} encryptedAnswer
     * @return {?}
     */
    verifyCaptcha(apiBaseUrl, nonce, answer, encryptedAnswer) {
        return this.httpClient
            .post(apiBaseUrl + '/verify/captcha', { nonce: nonce, answer: answer, validation: encryptedAnswer }, { observe: 'response' });
    }
    /**
     * @param {?} apiBaseUrl
     * @param {?} validation
     * @param {?=} translation
     * @return {?}
     */
    fetchAudio(apiBaseUrl, validation, translation) {
        /** @type {?} */
        const payload = { validation: validation };
        if (translation) {
            payload.translation = translation;
        }
        return this.httpClient
            .post(apiBaseUrl + '/captcha/audio', payload, { observe: 'response' });
    }
}
CaptchaDataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CaptchaDataService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CaptchaDataService.prototype.httpClient;
}
