// moh-common-lib/esm5/lib/components/thumbnail/thumbnail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, Output, EventEmitter, ViewContainerRef } from '@angular/core';
import { ModalDirective } from 'ngx-bootstrap';
import { CommonImage } from '../../../images/src/images';
var ThumbnailComponent = /** @class */ (function () {
    function ThumbnailComponent(viewContainerRef) {
        this.reviewMode = false;
        this.deleteImage = new EventEmitter();
        this.scaledWidth = 300;
        this.viewContainerRef = viewContainerRef;
    }
    /**
     * @return {?}
     */
    ThumbnailComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var scaledWidthString = (180 * this.imageObject.naturalWidth / this.imageObject.naturalHeight).toFixed(0);
        // console.log('scaled width: ' + scaledWidthString);
        this.scaledWidth = parseInt(scaledWidthString, 10);
        if (this.scaledWidth > 250) {
            // console.log('Scaled width > 250, drop it down to 250');
            this.scaledWidth = 250;
        }
        else if (this.scaledWidth < 30) {
            this.scaledWidth = 100;
        }
        if (isNaN(this.scaledWidth)) {
            this.scaledWidth = 300;
        }
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    ThumbnailComponent.prototype.delete = /**
     * @param {?} evt
     * @return {?}
     */
    function (evt) {
        console.log('ThumbnailComponent:Delete from thumbnail: %o', evt);
        console.log('ThumbnailComponent:imageObject:', JSON.stringify(this.imageObject, null, 2));
        this.deleteImage.emit(this.imageObject);
    };
    /**
     * @return {?}
     */
    ThumbnailComponent.prototype.showFullSizeView = /**
     * @return {?}
     */
    function () {
        this.fullSizeViewModal.config.backdrop = false;
        this.fullSizeViewModal.show();
    };
    /**
     * @return {?}
     */
    ThumbnailComponent.prototype.hideFullSizeView = /**
     * @return {?}
     */
    function () {
        this.fullSizeViewModal.hide();
    };
    ThumbnailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-thumbnail',
                    template: "<div class=\"thumbnail-container\">\n  <img [src]=\"imageObject.fileContent\" alt=\"\" tabindex=\"0\"\n    [ngClass]=\"{'image-thumbnail': scaledWidth <= 300, 'image-thumbnail-width-priority': scaledWidth > 300}\"\n    (click)=\"showFullSizeView()\">\n  <div class=\"action-strip\" *ngIf=\"!reviewMode\">\n    <a href=\"javascript:void(0)\" (click)=\"delete($event)\" class='text-danger' tabindex=\"0\">Remove</a>\n  </div>\n</div>\n\n<div bsModal #fullSizeViewModal=\"bs-modal\" class=\"modal fade\" tabindex=\"-1\" role=\"dialog\" aria-labelledby=\"myLargeModalLabel\" aria-hidden=\"true\">\n  <div class=\"modal-dialog modal-lg\">\n    <div class=\"modal-content\">\n      <div class=\"modal-header\">\n        <h4 class=\"modal-title\">{{imageObject.name}}</h4>\n        <button type=\"button\" class=\"close\" (click)=\"hideFullSizeView()\" aria-label=\"Close\">\n          <span aria-hidden=\"true\">&times;</span>\n        </button>\n      </div>\n      <div class=\"modal-body\">\n        <div class=\"row\">\n          <div class=\"col-md-12\" >\n            <img [src]=\"imageObject.fileContent\" alt=\"{{imageObject.name}}\" class=\"center-block img-fluid\">\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>",
                    styles: [".thumbnail-container{transition:.3s;-webkit-transform:translateY(0);transform:translateY(0)}.thumbnail-container .image-thumbnail,.thumbnail-container .image-thumbnail-width-priority{padding:2px 2px 0;border-radius:5px 5px 0 0}.thumbnail-container .image-thumbnail-width-priority:hover,.thumbnail-container .image-thumbnail:hover{cursor:zoom-in}.thumbnail-container .image-thumbnail{max-height:100px;height:auto;max-width:100%}.thumbnail-container .image-thumbnail-width-priority{max-width:270px;width:auto;max-height:100%}.thumbnail-container:hover{box-shadow:0 15px 10px -10px rgba(0,0,0,.1);-webkit-transform:translateY(-5px);transform:translateY(-5px)}.thumbnail-container:hover .action-strip{background:#ececec}.thumbnail-container .action-strip{height:2em;border-radius:0 0 5px 5px;text-align:right;font-size:small;margin:0 2px;color:#a94442;padding:.3em;transition:.3s}.thumbnail-container .action-strip a{text-decoration:none}"]
                }] }
    ];
    /** @nocollapse */
    ThumbnailComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    ThumbnailComponent.propDecorators = {
        imageObject: [{ type: Input }],
        reviewMode: [{ type: Input }],
        deleteImage: [{ type: Output }],
        fullSizeViewModal: [{ type: ViewChild, args: ['fullSizeViewModal',] }]
    };
    return ThumbnailComponent;
}());
export { ThumbnailComponent };
if (false) {
    /** @type {?} */
    ThumbnailComponent.prototype.imageObject;
    /** @type {?} */
    ThumbnailComponent.prototype.reviewMode;
    /** @type {?} */
    ThumbnailComponent.prototype.deleteImage;
    /** @type {?} */
    ThumbnailComponent.prototype.fullSizeViewModal;
    /**
     * @type {?}
     * @private
     */
    ThumbnailComponent.prototype.viewContainerRef;
    /** @type {?} */
    ThumbnailComponent.prototype.scaledWidth;
}
