// moh-common-lib/esm5/models/src/container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function WizardProgressItem() { }
if (false) {
    /** @type {?} */
    WizardProgressItem.prototype.title;
    /** @type {?} */
    WizardProgressItem.prototype.route;
}
/**
 * Base functionality for container that is used to display bread crumbs
 */
var /**
 * Base functionality for container that is used to display bread crumbs
 */
Container = /** @class */ (function () {
    function Container() {
    }
    /**
     * Converts a lower case string of a route in a user readable title.  e.g.
     * "personal-info" -> "Personal Info"
     */
    /**
     * Converts a lower case string of a route in a user readable title.  e.g.
     * "personal-info" -> "Personal Info"
     * @param {?} routePath
     * @return {?}
     */
    Container.prototype.convertRouteToTitle = /**
     * Converts a lower case string of a route in a user readable title.  e.g.
     * "personal-info" -> "Personal Info"
     * @param {?} routePath
     * @return {?}
     */
    function (routePath) {
        return routePath.split('-').map((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x[0].toUpperCase() + x.slice(1); })).join(' ');
    };
    /**
     * @protected
     * @param {?} pageRoutes
     * @return {?}
     */
    Container.prototype.setProgressSteps = /**
     * @protected
     * @param {?} pageRoutes
     * @return {?}
     */
    function (pageRoutes) {
        var _this = this;
        // Interface for wizard progress items
        this.progressSteps = pageRoutes.map((/**
         * @param {?} page
         * @return {?}
         */
        function (page) {
            if (page.path !== '') {
                return {
                    title: _this.convertRouteToTitle(page.path),
                    route: page.path
                };
            }
        })).filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x; }));
    };
    return Container;
}());
/**
 * Base functionality for container that is used to display bread crumbs
 */
export { Container };
if (false) {
    /**
     * Route items for the stepper
     * @type {?}
     */
    Container.prototype.progressSteps;
}
