// moh-common-lib/esm5/models/src/mask.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// List of constants used for masking inputs
import { EventEmitter, Input, Output } from '@angular/core';
import { Base } from './base';
/** @type {?} */
export var LETTER = /[A-Z]/i;
// Ignore case here, then upperCase it via pipe.
/** @type {?} */
export var NUMBER = /\d/;
/** @type {?} */
export var SPACE = ' ';
var MaskModel = /** @class */ (function (_super) {
    tslib_1.__extends(MaskModel, _super);
    function MaskModel() {
        var _this = _super.call(this) || this;
        _this.valueChange = new EventEmitter();
        _this.disabled = false;
        _this.required = true;
        return _this;
    }
    /**
     * Upper cases letters in string
     */
    /**
     * Upper cases letters in string
     * @param {?} text
     * @return {?}
     */
    MaskModel.prototype.upperCasePipe = /**
     * Upper cases letters in string
     * @param {?} text
     * @return {?}
     */
    function (text) {
        return text.toUpperCase();
    };
    /**
     * Updates the value
     */
    /**
     * Updates the value
     * @param {?} value
     * @return {?}
     */
    MaskModel.prototype.onUpdate = /**
     * Updates the value
     * @param {?} value
     * @return {?}
     */
    function (value) {
        // Emit value without spaces
        this.valueChange.emit(value);
    };
    MaskModel.propDecorators = {
        value: [{ type: Input }],
        valueChange: [{ type: Output }],
        disabled: [{ type: Input }],
        required: [{ type: Input }]
    };
    return MaskModel;
}(Base));
export { MaskModel };
if (false) {
    /** @type {?} */
    MaskModel.prototype.value;
    /** @type {?} */
    MaskModel.prototype.valueChange;
    /** @type {?} */
    MaskModel.prototype.disabled;
    /** @type {?} */
    MaskModel.prototype.required;
    /** @type {?} */
    MaskModel.prototype.mask;
    /** @type {?} */
    MaskModel.prototype.placeholder;
}
