// moh-common-lib/esm2015/lib/components/page-section/page-section.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
export class PageSectionComponent {
    constructor() {
        this.layout = 'tips';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
PageSectionComponent.decorators = [
    { type: Component, args: [{
                selector: 'common-page-section',
                template: "<div class=\"row\" [ngSwitch]=\"layout\">\n\n  <ng-container *ngSwitchCase=\"'tips'\">\n    <div class=\"col-md-8\">\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n    </div>\n    <div class=\"col-md-4 aside-col\">\n      <ng-container *ngTemplateOutlet=\"aside\"></ng-container>\n    </div>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"'double'\">\n    <div class=\"col-md-6\">\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n    </div>\n    <div class=\"col-md-6\">\n      <ng-container *ngTemplateOutlet=\"aside\"></ng-container>\n    </div>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"'noTips'\">\n      <div class=\"col-md-12\">\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n    </div>\n  </ng-container>\n\n</div>\n\n<!-- We use ng-template here to get around a bug with having multiple ng-contents in one template. By default, if there are duplicate ng-contents in a template Angular will select the very first one - even if latter ones are 'removed' by ngSwitch or ngIf.-->\n<ng-template #content>\n  <ng-content></ng-content>\n</ng-template>\n\n<ng-template #aside>\n  <ng-content select='aside'></ng-content>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                styles: ["common-page-section{display:block}.h5,h5{font-size:1rem;font-weight:700;margin-bottom:0}p+.h5,p+h5{margin-top:1.5rem}.aside-col aside{background:#f2f2f2;padding:1em;border-radius:5px}@media (min-width:768px){.aside-col aside{margin-top:1rem}}.aside-col aside ul{padding-left:0;list-style-type:none}.aside-col aside ul>li:before{content:\" - \";text-indent:-5px}"]
            }] }
];
/** @nocollapse */
PageSectionComponent.ctorParameters = () => [];
PageSectionComponent.propDecorators = {
    layout: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PageSectionComponent.prototype.layout;
}
