// moh-common-lib/esm2015/images/src/images.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UUID } from 'angular2-uuid';
/** @enum {number} */
const CommonImageError = {
    WrongType: 0,
    TooSmall: 1,
    TooBig: 2,
    AlreadyExists: 3,
    Unknown: 4,
    CannotOpen: 5,
    PDFnotSupported: 6,
    CannotOpenPDF: 7,
};
export { CommonImageError };
CommonImageError[CommonImageError.WrongType] = 'WrongType';
CommonImageError[CommonImageError.TooSmall] = 'TooSmall';
CommonImageError[CommonImageError.TooBig] = 'TooBig';
CommonImageError[CommonImageError.AlreadyExists] = 'AlreadyExists';
CommonImageError[CommonImageError.Unknown] = 'Unknown';
CommonImageError[CommonImageError.CannotOpen] = 'CannotOpen';
CommonImageError[CommonImageError.PDFnotSupported] = 'PDFnotSupported';
CommonImageError[CommonImageError.CannotOpenPDF] = 'CannotOpenPDF';
export class CommonImageProcessingError {
    // added errorDescription.PDF.JS gives proper error messages as invalid pdf structure or password protected pdf.Good for splunk tracking
    /**
     * @param {?} errorCode
     * @param {?=} errorDescription
     */
    constructor(errorCode, errorDescription) {
        this.errorCode = errorCode;
        this.errorDescription = errorDescription;
    }
}
if (false) {
    /** @type {?} */
    CommonImageProcessingError.prototype.commonImage;
    /** @type {?} */
    CommonImageProcessingError.prototype.rawImageFile;
    /** @type {?} */
    CommonImageProcessingError.prototype.maxSizeAllowed;
    /** @type {?} */
    CommonImageProcessingError.prototype.errorCode;
    /** @type {?} */
    CommonImageProcessingError.prototype.errorDescription;
}
/**
 * Image as uploaded by user
 */
export class CommonImage {
    constructor() {
        this.attachmentOrder = 0;
        this.uuid = UUID.UUID();
    }
}
if (false) {
    /** @type {?} */
    CommonImage.prototype.uuid;
    /** @type {?} */
    CommonImage.prototype.fileContent;
    /** @type {?} */
    CommonImage.prototype.contentType;
    /** @type {?} */
    CommonImage.prototype.size;
    /** @type {?} */
    CommonImage.prototype.sizeUnit;
    /** @type {?} */
    CommonImage.prototype.sizeTxt;
    /** @type {?} */
    CommonImage.prototype.naturalHeight;
    /** @type {?} */
    CommonImage.prototype.naturalWidth;
    /** @type {?} */
    CommonImage.prototype.name;
    /** @type {?} */
    CommonImage.prototype.id;
    /** @type {?} */
    CommonImage.prototype.error;
    /** @type {?} */
    CommonImage.prototype.attachmentOrder;
}
/**
 * @record
 */
export function CommonImageScaleFactors() { }
if (false) {
    /** @type {?} */
    CommonImageScaleFactors.prototype.widthFactor;
    /** @type {?} */
    CommonImageScaleFactors.prototype.heightFactor;
    /**
     * @param {?} scale
     * @return {?}
     */
    CommonImageScaleFactors.prototype.scaleDown = function (scale) { };
}
export class CommonImageScaleFactorsImpl {
    /**
     * @param {?} wFactor
     * @param {?} hFactor
     */
    constructor(wFactor, hFactor) {
        this.widthFactor = wFactor;
        this.heightFactor = hFactor;
    }
    /**
     * @param {?} scale
     * @return {?}
     */
    scaleDown(scale) {
        return new CommonImageScaleFactorsImpl(this.widthFactor * scale, this.heightFactor * scale);
    }
}
if (false) {
    /** @type {?} */
    CommonImageScaleFactorsImpl.prototype.widthFactor;
    /** @type {?} */
    CommonImageScaleFactorsImpl.prototype.heightFactor;
}
