// moh-common-lib/esm5/lib/components/form-action-bar/form-action-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, forwardRef } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
var FormActionBarComponent = /** @class */ (function () {
    function FormActionBarComponent() {
        this.submitLabel = 'Continue';
        this.canContinue = true;
        this.isLoading = false;
        this.defaultColor = true;
        this.btnClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FormActionBarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    FormActionBarComponent.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this.isLoading && this.canContinue) {
            this.btnClick.emit($event);
        }
        $event.stopPropagation();
        return false;
    };
    FormActionBarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-form-action-bar',
                    template: "<div class=\"form-action-bar form-bar\" [ngClass]=\"{disabled: !canContinue}\">\n    <button class=\"btn btn-lg {{defaultColor ? 'btn-primary' : 'btn-secondary' }} submit\"\n            (click)=\"onClick($event)\"\n            [ngClass]=\"{disabled: !canContinue || isLoading}\">\n      <ng-container *ngIf=\"!isLoading; else loadingSpinner\"> {{submitLabel}} </ng-container>\n      \n    </button>\n</div>\n\n<ng-template #loadingSpinner>\n    <i class=\"fa fa-spinner fa-pulse fa-fw\"></i>\n</ng-template>",
                    viewProviders: [{ provide: ControlContainer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return NgForm; })) }],
                    styles: [".form-action-bar{display:flex;justify-content:flex-end;padding:1em;background-color:#cdd9e4;transition:background-color .3s;position:-webkit-sticky;position:sticky;bottom:0;z-index:10;left:5rem;right:5rem}.form-action-bar.disabled{background-color:#d5d9dd}@media (max-width:991.98px){.form-action-bar{left:2.5rem;right:2.5rem}}.submit{min-width:240px}@media (max-width:767.98px){.form-action-bar{left:0;right:0}.submit{min-width:100%}}.btn{white-space:normal!important;word-wrap:break-word}"]
                }] }
    ];
    /** @nocollapse */
    FormActionBarComponent.ctorParameters = function () { return []; };
    FormActionBarComponent.propDecorators = {
        submitLabel: [{ type: Input }],
        canContinue: [{ type: Input }],
        isLoading: [{ type: Input }],
        defaultColor: [{ type: Input }],
        btnClick: [{ type: Output }]
    };
    return FormActionBarComponent;
}());
export { FormActionBarComponent };
if (false) {
    /** @type {?} */
    FormActionBarComponent.prototype.submitLabel;
    /** @type {?} */
    FormActionBarComponent.prototype.canContinue;
    /** @type {?} */
    FormActionBarComponent.prototype.isLoading;
    /** @type {?} */
    FormActionBarComponent.prototype.defaultColor;
    /** @type {?} */
    FormActionBarComponent.prototype.btnClick;
}
