// moh-common-lib/esm5/lib/components/file-uploader/file-uploader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, NgZone, Output, ViewChild, forwardRef } from '@angular/core';
import { NgForm, ControlContainer } from '@angular/forms';
import { Observable, fromEvent, merge } from 'rxjs';
import { map, filter, flatMap, scan, delay, retryWhen } from 'rxjs/operators';
import { CommonImage, CommonImageError, CommonImageProcessingError, CommonImageScaleFactorsImpl } from '../../../images/src/public_api';
// import { MspLogService } from '../../service/log.service';
// import { MspDataService } from '../../service/msp-data.service';
// import { BaseComponent } from '../base.component';
// import { LogEntry } from '../logging/log-entry.model';
// import {Person} from '../../model/application.model';
import { Router } from '@angular/router';
import { Base } from '../../../models/src/base';
// import {ApplicationBase} from '../../model/application-base.model';
// const loadImage = require('blueimp-load-image');
// const sha1 = require('sha1');
import * as loadImage_ from 'blueimp-load-image';
/** @type {?} */
var loadImage = loadImage_;
import * as sha1_ from 'sha1';
/** @type {?} */
var sha1 = sha1_;
// const PDFJS: PDFJSStatic = require('pdfjs-dist');
import * as PDFJS_ from 'pdfjs-dist';
/** @type {?} */
var PDFJS = ((/** @type {?} */ (PDFJS_)));
/**
 * @record
 */
export function FileUploaderMsg() { }
if (false) {
    /** @type {?} */
    FileUploaderMsg.prototype.required;
}
// TODO - Remove this and fix tslint issues
/* tslint:disable:max-line-length*/
var FileUploaderComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FileUploaderComponent, _super);
    function FileUploaderComponent(zone, cd, router, controlContainer) {
        var _this = _super.call(this) || this;
        _this.zone = zone;
        _this.cd = cd;
        _this.router = router;
        _this.controlContainer = controlContainer;
        // lang = require('./i18n');
        _this.noIdImage = false;
        // @ContentChild('uploadInstruction') uploadInstructionRef: ElementRef;
        _this.images = new Array(0);
        _this.imagesChange = new EventEmitter();
        _this.required = false;
        _this.instructionText = 'Please upload required ID documents.';
        _this.errorMessages = { required: 'File is required.' };
        _this.errorDocument = new EventEmitter();
        return _this;
        // this.application = this.getApplicationType();
    }
    Object.defineProperty(FileUploaderComponent.prototype, "fileControl", {
        /**
         * This is created as a workaround to access the form control that binds to
         * the input[type='file']. We can't access it via the template name bindings
         * as that isn't working, so instead we access the parent form and then find
         * the input by name.
         */
        get: /**
         * This is created as a workaround to access the form control that binds to
         * the input[type='file']. We can't access it via the template name bindings
         * as that isn't working, so instead we access the parent form and then find
         * the input by name.
         * @return {?}
         */
        function () {
            /** @type {?} */
            var INPUT_NAME = "fileUploadBrowse-" + this.id;
            // note - should be "this.controlContainer as NgForm" here, which works,
            // but fails on compiliation due to secondary entries
            return ((/** @type {?} */ (this.controlContainer))).controls[INPUT_NAME];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Return true if file already exists in the list; false otherwise.
     */
    /**
     * Return true if file already exists in the list; false otherwise.
     * @param {?} file
     * @param {?} imageList
     * @return {?}
     */
    FileUploaderComponent.checkImageExists = /**
     * Return true if file already exists in the list; false otherwise.
     * @param {?} file
     * @param {?} imageList
     * @return {?}
     */
    function (file, imageList) {
        if (!imageList || imageList.length < 1) {
            return false;
        }
        else {
            /** @type {?} */
            var sha1Sum = sha1(file.fileContent);
            for (var i = imageList.length - 1; i >= 0; i--) {
                // console.log(`compare  ${imageList[i].id} with ${sha1Sum}, result ${imageList[i].id === sha1Sum}`);
                if (imageList[i].id === sha1Sum) {
                    console.log("This file " + file.name + " has already been uploaded.");
                    return true;
                }
            }
            return false;
        }
    };
    /**
     * A special method to force the rendering of this component.  This is a workaround
     * because for some unknown reason, AngularJS2 change detector does not detect the
     * change of the images Array.
     */
    /**
     * A special method to force the rendering of this component.  This is a workaround
     * because for some unknown reason, AngularJS2 change detector does not detect the
     * change of the images Array.
     * @return {?}
     */
    FileUploaderComponent.prototype.forceRender = /**
     * A special method to force the rendering of this component.  This is a workaround
     * because for some unknown reason, AngularJS2 change detector does not detect the
     * change of the images Array.
     * @return {?}
     */
    function () {
        this.zone.run((/**
         * @return {?}
         */
        function () {
        }));
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FileUploaderComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        console.log('fileuploader onChanges', changes['images']);
        // if (changes['images'] && (changes['images'].currentValue.length === 0 &&
        //   changes['images'].previousValue.length > 0)) {
        if (changes['images'] && (changes['images'].currentValue.length === 0
            && changes['images'].previousValue
            && changes['images'].previousValue.length > 0)) {
            this.noIdImage = true;
        }
        else {
            this.noIdImage = false;
        }
    };
    /*
     System processing steps

     1. User clicks browse or drag-n-drops an file
     2. For browse case, the browser is told to only accept mime type image/*, .JPG, .GIF, .PNG, etc,
     however user can override and for drag-n-drop we don't can't impose this filter
     4. Using the HTML5 File API, we open a handle on the file
     5. Read the filename for later display to the user
     6. Create a hidden Image element in the browser's DOM
     7. Read the file's bytes as a DataUrl and copy them into the Image element
     8. Wait until the Image finishes loading the image
     9. Read the image element's natural width and height
     10. Pass the File handle into a HTML5 Canvas lib (we need the XIFF headers to auto rotate, XIFF headers are not available in DataUrl)
     11. The Canvas errors because it's a wrong type, e.g., TIFF, we abort and notify user
     12. Instruct the Canvas lib to keep resizing the image if it exceeds a maximum width or height,
     extract meta data, and auto-orient based on XIFF metadata.  It uses a "contain" operation which retains
     it's width to height pixel ratio.
     13. Call a function on the Canvas element to turn the Canvas into a JPEG of quality 50%.
     14. Once in a Blob with get the blob size in bytes and a human friendly display size
     15. In order to more easily manage the image, we convert the Blob to a DataUrl again.
     16. Pass the DataUrl into a hash algorithm to create an identifier and to check if the image has already been uploaded
     17. Next we check the final size of the image to ensure it's not to small in resolution
     (arguably this could've been done earlier), if too small we notify user
     18. Finally, the image is saved into the user's ongoing EA/PA application including localstorage
     19. The image is displayed to user as a thumbnail

     */
    /*
         System processing steps
    
         1. User clicks browse or drag-n-drops an file
         2. For browse case, the browser is told to only accept mime type image/*, .JPG, .GIF, .PNG, etc,
         however user can override and for drag-n-drop we don't can't impose this filter
         4. Using the HTML5 File API, we open a handle on the file
         5. Read the filename for later display to the user
         6. Create a hidden Image element in the browser's DOM
         7. Read the file's bytes as a DataUrl and copy them into the Image element
         8. Wait until the Image finishes loading the image
         9. Read the image element's natural width and height
         10. Pass the File handle into a HTML5 Canvas lib (we need the XIFF headers to auto rotate, XIFF headers are not available in DataUrl)
         11. The Canvas errors because it's a wrong type, e.g., TIFF, we abort and notify user
         12. Instruct the Canvas lib to keep resizing the image if it exceeds a maximum width or height,
         extract meta data, and auto-orient based on XIFF metadata.  It uses a "contain" operation which retains
         it's width to height pixel ratio.
         13. Call a function on the Canvas element to turn the Canvas into a JPEG of quality 50%.
         14. Once in a Blob with get the blob size in bytes and a human friendly display size
         15. In order to more easily manage the image, we convert the Blob to a DataUrl again.
         16. Pass the DataUrl into a hash algorithm to create an identifier and to check if the image has already been uploaded
         17. Next we check the final size of the image to ensure it's not to small in resolution
         (arguably this could've been done earlier), if too small we notify user
         18. Finally, the image is saved into the user's ongoing EA/PA application including localstorage
         19. The image is displayed to user as a thumbnail
    
         */
    /**
     * @return {?}
     */
    FileUploaderComponent.prototype.ngOnInit = /*
         System processing steps
    
         1. User clicks browse or drag-n-drops an file
         2. For browse case, the browser is told to only accept mime type image/*, .JPG, .GIF, .PNG, etc,
         however user can override and for drag-n-drop we don't can't impose this filter
         4. Using the HTML5 File API, we open a handle on the file
         5. Read the filename for later display to the user
         6. Create a hidden Image element in the browser's DOM
         7. Read the file's bytes as a DataUrl and copy them into the Image element
         8. Wait until the Image finishes loading the image
         9. Read the image element's natural width and height
         10. Pass the File handle into a HTML5 Canvas lib (we need the XIFF headers to auto rotate, XIFF headers are not available in DataUrl)
         11. The Canvas errors because it's a wrong type, e.g., TIFF, we abort and notify user
         12. Instruct the Canvas lib to keep resizing the image if it exceeds a maximum width or height,
         extract meta data, and auto-orient based on XIFF metadata.  It uses a "contain" operation which retains
         it's width to height pixel ratio.
         13. Call a function on the Canvas element to turn the Canvas into a JPEG of quality 50%.
         14. Once in a Blob with get the blob size in bytes and a human friendly display size
         15. In order to more easily manage the image, we convert the Blob to a DataUrl again.
         16. Pass the DataUrl into a hash algorithm to create an identifier and to check if the image has already been uploaded
         17. Next we check the final size of the image to ensure it's not to small in resolution
         (arguably this could've been done earlier), if too small we notify user
         18. Finally, the image is saved into the user's ongoing EA/PA application including localstorage
         19. The image is displayed to user as a thumbnail
    
         */
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var dragOverStream = fromEvent(this.dropZone.nativeElement, 'dragover');
        /**
         * Must cancel the dragover event in order for the drop event to work.
         */
        dragOverStream.pipe(map((/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            return event;
        }))).subscribe((/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            // console.log('Cancel dragover event.');
            evt.preventDefault();
        }));
        /** @type {?} */
        var dropStream = fromEvent(this.dropZone.nativeElement, 'drop');
        /** @type {?} */
        var filesArrayFromDrop = dropStream.pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
            return event.dataTransfer.files;
        })));
        /** @type {?} */
        var browseFileStream = fromEvent(this.browseFileRef.nativeElement, 'change');
        // const captureFileStream = fromEvent<Event>(this.captureFileRef.nativeElement, 'change');
        merge(merge(browseFileStream).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
            return event.target['files'];
        }))), filesArrayFromDrop).pipe(filter((/**
         * @param {?} files
         * @return {?}
         */
        function (files) {
            console.log('files');
            return !!files && files.length && files.length > 0;
        })), flatMap((/**
         * @param {?} fileList
         * @return {?}
         */
        function (fileList) {
            return _this.observableFromFiles(fileList, new CommonImageScaleFactorsImpl(1, 1));
        })), filter((/**
         * @param {?} mspImage
         * @return {?}
         */
        function (mspImage) {
            /** @type {?} */
            var imageExists = FileUploaderComponent.checkImageExists(mspImage, _this.images);
            if (imageExists) {
                _this.handleError(CommonImageError.AlreadyExists, mspImage);
                _this.resetInputFields();
            }
            return !imageExists;
        })), 
        // TODO - Is this necessary? Can likely be removed as it's exactly identical to the preceding.
        filter((/**
         * @param {?} mspImage
         * @return {?}
         */
        function (mspImage) {
            /** @type {?} */
            var imageExists = FileUploaderComponent.checkImageExists(mspImage, _this.images);
            if (imageExists) {
                _this.handleError(CommonImageError.AlreadyExists, mspImage);
                _this.resetInputFields();
            }
            return !imageExists;
        })), filter((/**
         * @param {?} mspImage
         * @return {?}
         */
        function (mspImage) {
            /** @type {?} */
            var imageSizeOk = _this.checkImageDimensions(mspImage);
            if (!imageSizeOk) {
                _this.handleError(CommonImageError.TooSmall, mspImage);
                _this.resetInputFields();
            }
            return imageSizeOk;
        }))).subscribe((/**
         * @param {?} file
         * @return {?}
         */
        function (file) {
            _this.handleImageFile(file);
            _this.resetInputFields();
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.log('Error in loading image: %o', error);
            /**
             * Handle the error if the image is gigantic that after
             * 100 times of scaling down by 30% on each step, the image
             * is still over 1 MB.
             */
            if (error.errorCode) {
                if (CommonImageError.TooBig === error.errorCode) {
                    _this.handleError(CommonImageError.TooBig, error.image);
                }
                else if (CommonImageError.CannotOpen === error.errorCode) {
                    if (!error.image) {
                        error.image = new CommonImage();
                        if (error.rawImageFile) {
                            error.image.name = error.rawImageFile.name;
                        }
                    }
                    _this.handleError(CommonImageError.CannotOpen, error.image);
                }
                else if (CommonImageError.CannotOpenPDF === error.errorCode) {
                    _this.handleError(CommonImageError.CannotOpenPDF, error.image, error.errorDescription);
                }
                else {
                    throw error;
                }
            }
        }), (/**
         * @return {?}
         */
        function () {
            console.log('completed loading image');
        }));
    };
    /**
     * @param {?} var1
     * @return {?}
     */
    FileUploaderComponent.prototype.test = /**
     * @param {?} var1
     * @return {?}
     */
    function (var1) {
        console.log(var1);
    };
    /**
     * @return {?}
     */
    FileUploaderComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var imagePlaceholderEnterKeyStream = merge(fromEvent(this.imagePlaceholderRef.nativeElement, 'keyup'), fromEvent(this.selectFileLabelRef.nativeElement, 'keyup')).pipe(filter((/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) { return evt.key === 'Enter'; })));
        merge(fromEvent(this.imagePlaceholderRef.nativeElement, 'click'), 
        // fromEvent<Event>(this.uploadInstructionRef.nativeElement, 'click'),
        imagePlaceholderEnterKeyStream).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
            return event;
        }))).subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { _this.browseFileRef.nativeElement.click(); }));
    };
    /**
     * Solve size in this equation: size * 0.8to-the-power-of30 < 1MB, size
     * will be the max image size this application can accept and scale down
     * to under 1MB. In this case: size < 807 MB
     *
     * 30 is the number of retries. the value for maxRetry passed to retryStrategy
     * function.
     *
     * If: size * 0.8to-the-power-of40 < 1MB, then size < 1262 MB.
     *
     * Note: 0.8 is the self.appConstants.images.reductionScaleFactor defined in global.js
     *
     *
     * @param file
     * @param scaleFactors
     */
    /**
     * Solve size in this equation: size * 0.8to-the-power-of30 < 1MB, size
     * will be the max image size this application can accept and scale down
     * to under 1MB. In this case: size < 807 MB
     *
     * 30 is the number of retries. the value for maxRetry passed to retryStrategy
     * function.
     *
     * If: size * 0.8to-the-power-of40 < 1MB, then size < 1262 MB.
     *
     * Note: 0.8 is the self.appConstants.images.reductionScaleFactor defined in global.js
     *
     *
     * @param {?} fileList
     * @param {?} scaleFactors
     * @return {?}
     */
    FileUploaderComponent.prototype.observableFromFiles = /**
     * Solve size in this equation: size * 0.8to-the-power-of30 < 1MB, size
     * will be the max image size this application can accept and scale down
     * to under 1MB. In this case: size < 807 MB
     *
     * 30 is the number of retries. the value for maxRetry passed to retryStrategy
     * function.
     *
     * If: size * 0.8to-the-power-of40 < 1MB, then size < 1262 MB.
     *
     * Note: 0.8 is the self.appConstants.images.reductionScaleFactor defined in global.js
     *
     *
     * @param {?} fileList
     * @param {?} scaleFactors
     * @return {?}
     */
    function (fileList, scaleFactors) {
        var _this = this;
        /**
         * Previously this was set in appConstants, but that's removed from the common lib.
         * @type {?}
         */
        var reductionScaleFactor = 0.8;
        console.log('obserablveFromFiles');
        // Init
        /** @type {?} */
        var self = this;
        //    let  pageNumber = Math.max(...self.images.concat( self.application.getAllImages()).map(function(o) {return o.attachmentOrder; }), 0) + 1 ;
        /** @type {?} */
        var pageNumber = Math.max.apply(Math, tslib_1.__spread(self.images.map((/**
         * @param {?} o
         * @return {?}
         */
        function (o) { return o.attachmentOrder; })), [0])) + 1;
        // Create our observer
        /** @type {?} */
        var fileObservable = Observable.create((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            /** @type {?} */
            var mspImages = [];
            scaleFactors = scaleFactors.scaleDown(reductionScaleFactor);
            for (var fileIndex = 0; fileIndex < fileList.length; fileIndex++) {
                /** @type {?} */
                var file = fileList[fileIndex];
                console.log('Start processing file ' + fileIndex + ' of ' + fileList.length + ' %s of size %s bytes %s type', file.name, file.size, file.type);
                /* Previously set in appConstants */
                /** @type {?} */
                var pdfScaleFactor = 2.0;
                // let mspImage: MspImage = new MspImage();
                // let reader: FileReader = new FileReader();
                // // Copy file properties
                // mspImage.name = file.name;
                if (file.type === 'application/pdf') {
                    // this.logService.log({name: file.name + ' Received in Upload',
                    //     UUID: self.dataService.getMspUuid()}, 'File_Upload');
                    /**
                     *  Page number logic :
                     *      Images - Assign current page number whichever is available..so get the current page number , pass it to call back [reserve it] and increment
                     *      PDF    -  we dont know how many pages..so cant get current number and keep it since it can be multiple pages... so start assigning later point
                     *      when PDF is totally read..
                     *
                     *  */
                    _this.readPDF(file, pdfScaleFactor, (/**
                     * @param {?} images
                     * @param {?} pdfFile
                     * @return {?}
                     */
                    function (images, pdfFile) {
                        // this.logService.log({name: file.name + 'is successfully split into ' + images.length + ' images',
                        // UUID: self.dataService.getMspUuid()}, 'File_Upload');
                        images.map((/**
                         * @param {?} image
                         * @param {?} index
                         * @return {?}
                         */
                        function (image, index) {
                            image.name = pdfFile.name;
                            _this.resizeImage(image, self, scaleFactors, observer, pageNumber, true); // index starts from zero
                            pageNumber = pageNumber + 1;
                        }));
                    }), (/**
                     * @param {?} error
                     * @return {?}
                     */
                    function (error) {
                        console.log('error' + JSON.stringify(error));
                        /** @type {?} */
                        var imageReadError = new CommonImageProcessingError(CommonImageError.CannotOpenPDF, error);
                        observer.error(imageReadError);
                    }));
                }
                else {
                    // Load image into img element to read natural height and width
                    _this.readImage(file, pageNumber, (/**
                     * @param {?} image
                     * @param {?} imageFile
                     * @param {?} nextPageNumber
                     * @return {?}
                     */
                    function (image, imageFile, nextPageNumber) {
                        image.id = imageFile.name; // .name deprecated, changed image.name to image.id
                        _this.resizeImage(image, self, scaleFactors, observer, nextPageNumber);
                    }), (
                    // can be ignored for bug, the log line is never called
                    // can be ignored for bug, the log line is never called
                    /**
                     * @param {?} error
                     * @return {?}
                     */
                    function (error) {
                        console.log('error' + JSON.stringify(error));
                        observer.error(error);
                    }));
                    pageNumber = pageNumber + 1;
                }
            }
            // retryWhen is potential issue!
        })).pipe(retryWhen(this.retryStrategy(32)));
        return fileObservable;
    };
    /**
     * @private
     * @param {?} image
     * @param {?} self
     * @param {?} scaleFactors
     * @param {?} observer
     * @param {?=} pageNumber
     * @param {?=} isPdf
     * @return {?}
     */
    FileUploaderComponent.prototype.resizeImage = /**
     * @private
     * @param {?} image
     * @param {?} self
     * @param {?} scaleFactors
     * @param {?} observer
     * @param {?=} pageNumber
     * @param {?=} isPdf
     * @return {?}
     */
    function (image, self, scaleFactors, observer, pageNumber, isPdf) {
        if (pageNumber === void 0) { pageNumber = 0; }
        if (isPdf === void 0) { isPdf = false; }
        // While it's still in an image, get it's height and width
        /** @type {?} */
        var mspImage = new CommonImage();
        /** @type {?} */
        var reader = new FileReader();
        console.log('image.name:' + image.id); // .name deprecated, changed image.name to image.id
        // Copy file properties
        mspImage.name = image.id;
        if (isPdf) {
            mspImage.name = image.name + '-page' + pageNumber; // Just give name to pdf
        }
        // Temporary so we don't have duplicate file names. TODO: Improve.
        //   mspImage.name += Math.ceil(Math.random()*100);
        mspImage.attachmentOrder = pageNumber;
        mspImage.naturalWidth = image.naturalWidth;
        mspImage.naturalHeight = image.naturalHeight;
        console.log("image file natural height and width:\n            " + mspImage.naturalHeight + " x " + mspImage.naturalWidth);
        // Canvas will force the change to a JPEG
        mspImage.contentType = 'image/jpeg'; // previously in appConstants
        // Scale the image by loading into a canvas
        console.log('Start scaling down the image using blueimp-load-image lib: ');
        /** @type {?} */
        var scaledImage = loadImage(image.src, (
        // NOTE: we pass the File ref here again even though its already read because we need the XIFF metadata
        /**
         * @param {?} canvas
         * @param {?} metadata
         * @return {?}
         */
        function (canvas, metadata) {
            // Canvas may be an Event when errors happens
            if (canvas instanceof Event) {
                self.handleError(CommonImageError.WrongType, mspImage);
                self.resetInputFields();
                return;
            }
            // Convert to blob to get size
            canvas.toBlob((/**
             * @param {?} blob
             * @return {?}
             */
            function (blob) {
                // Copy the blob properties
                mspImage.size = blob.size;
                // log image info (but only for the first time before any scaling)
                // if (s
                /** @type {?} */
                var fileName = mspImage.name;
                /** @type {?} */
                var nBytes = mspImage.size;
                /** @type {?} */
                var fileSize = '';
                /** @type {?} */
                var fileSizeUnit = '';
                /** @type {?} */
                var sOutput = nBytes + ' bytes';
                // optional code for multiples approximation
                for (var aMultiples = ['KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'], nMultiple = 0, nApprox = nBytes / 1024; nApprox > 1; nApprox /= 1024, nMultiple++) {
                    sOutput = nApprox.toFixed(3) + ' ' + aMultiples[nMultiple] + ' (' + nBytes + ' bytes)';
                    fileSize = nApprox.toFixed(0);
                    fileSizeUnit = aMultiples[nMultiple];
                    mspImage.sizeUnit = fileSizeUnit;
                }
                console.log("File " + fileName + " is scaled down to: " + sOutput);
                mspImage.sizeTxt = sOutput;
                // call reader with new transformed image
                reader.onload = (/**
                 * @param {?} evt
                 * @return {?}
                 */
                function (evt) {
                    mspImage.fileContent = evt.target.result;
                    mspImage.id = sha1(mspImage.fileContent);
                    // keep scaling down the image until the image size is
                    // under max image size
                    /**
                     * previously in appConstants
                     * @type {?}
                     */
                    var maxSizeBytes = 1048576;
                    if (mspImage.size > maxSizeBytes) {
                        console.log('File size after scaling down: %d, max file size allowed: %d', mspImage.size, maxSizeBytes);
                        /** @type {?} */
                        var imageTooBigError = new CommonImageProcessingError(CommonImageError.TooBig);
                        imageTooBigError.maxSizeAllowed = maxSizeBytes;
                        imageTooBigError.commonImage = mspImage;
                        observer.error(imageTooBigError);
                    }
                    else {
                        // log image info
                        //   self.logImageInfo("msp_file-uploader_after_resize_attributes", self.dataService.getMspUuid(), mspImage);
                        observer.next(mspImage);
                    }
                });
                reader.readAsDataURL(blob);
            }), 
            // What mime type to make the blob as and jpeg quality
            'image/jpeg', 0.5);
        }), {
            maxWidth: 2600 * scaleFactors.widthFactor,
            maxHeight: 3300 * scaleFactors.heightFactor,
            contain: true,
            canvas: true,
            meta: true,
            orientation: true
        });
    };
    /**
     * Max retry scaling down for maxRetry times.
     */
    /**
     * Max retry scaling down for maxRetry times.
     * @param {?} maxRetry
     * @return {?}
     */
    FileUploaderComponent.prototype.retryStrategy = /**
     * Max retry scaling down for maxRetry times.
     * @param {?} maxRetry
     * @return {?}
     */
    function (maxRetry) {
        return (/**
         * @param {?} errors
         * @return {?}
         */
        function (errors) {
            /**Done: COMPLETE THIS! For some reason can't get scan() to work, types always malformed.*/
            // return errors.pipe(
            //     // scan((acc, curr) => {acc + curr}, 0)
            //     scan((acc, error, index) => {
            //         return acc + error;
            //     }, 0)
            // );
            // Done: Unsure if we have to re-implement this line. It causes errors, but simply removing it may not be appropriate.
            // NOTE: RxJS-compat might be saving us here and "fixing" the errors. See if errors return when we remove rxjs-compat.
            // return errors.pipe(scan((acc, curr) => acc + curr, 0))
            return errors.pipe(scan((
            // return errors.pipe(
            // return errors.pipe(
            /**
             * @param {?} acc
             * @param {?} error
             * @param {?} index
             * @return {?}
             */
            function (acc, error, index) {
                // console.log('Error encountered: %o', error);;
                /**
                 * If the error is about file too big and we have not reach max retry
                 * yet, theyt keep going to scaling down.
                 */
                if (acc < maxRetry && error.errorCode === CommonImageError.TooBig) {
                    // console.log('Progressively scaling down the image, step %d.', index);
                    return acc + 1;
                }
                else {
                    /**
                     * For either conditions terminate the retry, propogate
                     * the error.
                     *
                     * 1. errors such as CannotRead or any other unknown errors
                     * not listed in MspImageError enum
                     * 2. Exceeded maxRetry
                     *
                     */
                    console.log('Re-throw this image process error: %o', error);
                    throw error;
                }
            }), 0), delay(2));
        });
    };
    /**
     * @private
     * @param {?} imageFile
     * @param {?} nextPageNumber
     * @param {?} callback
     * @param {?} invalidImageHanlder
     * @return {?}
     */
    FileUploaderComponent.prototype.readImage = /**
     * @private
     * @param {?} imageFile
     * @param {?} nextPageNumber
     * @param {?} callback
     * @param {?} invalidImageHanlder
     * @return {?}
     */
    function (imageFile, nextPageNumber, callback, invalidImageHanlder) {
        /** @type {?} */
        var reader = new FileReader();
        reader.onload = (/**
         * @param {?} progressEvt
         * @return {?}
         */
        function (progressEvt) {
            console.log('loading image into an img tag: %o', progressEvt);
            // Load into an image element
            /** @type {?} */
            var imgEl = document.createElement('img');
            imgEl.src = ((/** @type {?} */ (reader.result)));
            // Wait for onload so all properties are populated
            imgEl.onload = (/**
             * @param {?} args
             * @return {?}
             */
            function (args) {
                console.log('Completed image loading into an img tag: %o', args);
                return callback(imgEl, imageFile, nextPageNumber);
            });
            imgEl.onerror =
                (/**
                 * @param {?} args
                 * @return {?}
                 */
                function (args) {
                    // log it to the console
                    console.log('This image cannot be opened/read, it is probably an invalid image. %o', args);
                    // throw new Error('This image cannot be opened/read');
                    /** @type {?} */
                    var imageReadError = new CommonImageProcessingError(CommonImageError.CannotOpen);
                    imageReadError.rawImageFile = imageFile;
                    return invalidImageHanlder(imageReadError);
                });
        });
        reader.readAsDataURL(imageFile);
    };
    /**
     * @private
     * @param {?} pdfFile
     * @param {?} pdfScaleFactor
     * @param {?} callback
     * @param {?} error
     * @return {?}
     */
    FileUploaderComponent.prototype.readPDF = /**
     * @private
     * @param {?} pdfFile
     * @param {?} pdfScaleFactor
     * @param {?} callback
     * @param {?} error
     * @return {?}
     */
    function (pdfFile, pdfScaleFactor, callback, error) {
        PDFJS.disableWorker = true;
        PDFJS.disableStream = true;
        /** @type {?} */
        var reader = new FileReader();
        /** @type {?} */
        var currentPage = 1;
        /** @type {?} */
        var canvas = document.createElement('canvas');
        /** @type {?} */
        var imgElsArray = [];
        /** @type {?} */
        var ctx = canvas.getContext('2d');
        reader.onload = (/**
         * @param {?} progressEvt
         * @return {?}
         */
        function (progressEvt) {
            /** @type {?} */
            var docInitParams = { data: reader.result };
            // TODO - The 'as any' was added when porting to common library from MSP
            PDFJS.getDocument(((/** @type {?} */ (docInitParams)))).then((/**
             * @param {?} pdfdoc
             * @return {?}
             */
            function (pdfdoc) {
                /** @type {?} */
                var numPages = pdfdoc.numPages;
                if (currentPage <= pdfdoc.numPages) {
                    getPage();
                }
                /**
                 * @return {?}
                 */
                function getPage() {
                    pdfdoc.getPage(currentPage).then((/**
                     * @param {?} page
                     * @return {?}
                     */
                    function (page) {
                        /** @type {?} */
                        var viewport = page.getViewport(pdfScaleFactor);
                        canvas.height = viewport.height;
                        canvas.width = viewport.width;
                        /** @type {?} */
                        var renderContext = {
                            canvasContext: ctx,
                            viewport: viewport
                        };
                        page.render(renderContext).then((/**
                         * @return {?}
                         */
                        function () {
                            /** @type {?} */
                            var imgEl = document.createElement('img');
                            imgEl.src = canvas.toDataURL();
                            imgElsArray.push(imgEl);
                            if (currentPage < numPages) {
                                currentPage++;
                                getPage();
                            }
                            else {
                                callback(imgElsArray, pdfFile);
                            }
                        }));
                    }), (/**
                     * @param {?} errorReason
                     * @return {?}
                     */
                    function (errorReason) {
                        error(errorReason);
                    }));
                }
            }), (/**
             * @param {?} errorReason
             * @return {?}
             */
            function (errorReason) {
                error(errorReason);
            }));
        });
        reader.readAsArrayBuffer(pdfFile);
    };
    /**
     * Non reversible image filter to take an existing canvas and make it gray scale
     * @param canvas
     */
    /**
     * Non reversible image filter to take an existing canvas and make it gray scale
     * @param {?} canvas
     * @return {?}
     */
    FileUploaderComponent.prototype.makeGrayScale = /**
     * Non reversible image filter to take an existing canvas and make it gray scale
     * @param {?} canvas
     * @return {?}
     */
    function (canvas) {
        /** @type {?} */
        var context = canvas.getContext('2d');
        /** @type {?} */
        var imageData = context.getImageData(0, 0, canvas.width, canvas.height);
        /** @type {?} */
        var data = imageData.data;
        for (var i = 0; i < data.length; i += 4) {
            /** @type {?} */
            var brightness = 0.34 * data[i] + 0.5 * data[i + 1] + 0.16 * data[i + 2];
            // red
            data[i] = brightness;
            // green
            data[i + 1] = brightness;
            // blue
            data[i + 2] = brightness;
        }
        // overwrite original image
        context.putImageData(imageData, 0, 0);
    };
    /**
     * @param {?} mspImage
     * @return {?}
     */
    FileUploaderComponent.prototype.handleImageFile = /**
     * @param {?} mspImage
     * @return {?}
     */
    function (mspImage) {
        console.log('image size (bytes) after compression: ' + mspImage.size);
        if (this.images.length >= 50) {
            // log it
            // this.logImageInfo('msp_file-uploader_error', this.dataService.getMspUuid(),
            //     mspImage, `Number of image files exceeds max of ${50}`);
            // log to console
            console.log("Max number of image file you can upload is " + 50 + ".\n      This file " + mspImage.name + " was not uploaded.");
        }
        else {
            this.images.push(mspImage);
            this.imagesChange.emit(this.images);
            this.showError = false;
            this.noIdImage = false;
        }
    };
    /**
     * @param {?} error
     * @param {?} mspImage
     * @param {?=} errorDescription
     * @return {?}
     */
    FileUploaderComponent.prototype.handleError = /**
     * @param {?} error
     * @param {?} mspImage
     * @param {?=} errorDescription
     * @return {?}
     */
    function (error, mspImage, errorDescription) {
        if (!mspImage) {
            mspImage = new CommonImage();
        }
        // just add the error to mspImage
        mspImage.error = error;
        // log the error
        if (error !== CommonImageError.PDFnotSupported) {
            // this.logImageInfo('msp_file-uploader_error', this.dataService.getMspUuid(), mspImage,
            //     '  mspImageFile: ' + mspImage.name + '  mspErrorNum: ' + error + '  mspError: ' +
            //     error + '-' + errorDescription);
        }
        // console.log("error with image: ", mspImage);
        this.errorDocument.emit(mspImage);
    };
    /**
     * Reset input fields so that user can delete a file and
     * immediately upload that file again.
     */
    /**
     * Reset input fields so that user can delete a file and
     * immediately upload that file again.
     * @return {?}
     */
    FileUploaderComponent.prototype.resetInputFields = /**
     * Reset input fields so that user can delete a file and
     * immediately upload that file again.
     * @return {?}
     */
    function () {
        // let brosweFileInputElement = this.browseFileRef.nativeElement;
        // let captureFileInputElement = this.captureFileRef.nativeElement;
        this.browseFileRef.nativeElement.value = '';
        // this.captureFileRef.nativeElement.value = '';
    };
    /**
     * @param {?} mspImage
     * @return {?}
     */
    FileUploaderComponent.prototype.deleteImage = /**
     * @param {?} mspImage
     * @return {?}
     */
    function (mspImage) {
        this.resetInputFields();
        this.images = this.images.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.uuid !== mspImage.uuid; }));
        this.imagesChange.emit(this.images);
        // If there are no images yet, we have to reset the input so it triggers 'required'.
        if (this.required && this.images.length <= 0) {
            console.log('No images, resetting input');
            // this.fileControl.value = '';
            this.fileControl.setErrors({ 'required': true });
        }
    };
    /**
     * Log image attributes
     * @param mspImage
     */
    /**
     * Log image attributes
     * @private
     * @param {?} title
     * @param {?} applicationId
     * @param {?} mspImage
     * @param {?=} additionalInfo
     * @return {?}
     */
    FileUploaderComponent.prototype.logImageInfo = /**
     * Log image attributes
     * @private
     * @param {?} title
     * @param {?} applicationId
     * @param {?} mspImage
     * @param {?=} additionalInfo
     * @return {?}
     */
    function (title, applicationId, mspImage, additionalInfo) {
        // TODO!
        // // create log entry
        // const log: LogEntry = new LogEntry();
        // log.applicationId = applicationId;
        // const now = moment();
        // log.mspTimestamp = now.toISOString();
        // log.applicationPhase = title + ':  mspImageId: ' + mspImage.id
        //     + '  mspImageUuid: ' + mspImage.uuid
        //     + '  mspImageSize: ' + mspImage.size
        //     + '  mspImageWidth: ' + mspImage.naturalWidth
        //     + '  mspImageHeight: ' + mspImage.naturalHeight
        //     + '  mspImageContentType: ' + mspImage.contentType
        //     + (additionalInfo ? '  ' + additionalInfo : '');
        // // send it while subscribing to response
        // this.logService.logIt(log, title).subscribe(
        //     (response) => {
        //         // console.log('log rest service response: ');
        //         // console.log(response);
        //     },
        //     (error) => {
        //         console.log('HTTP error response from logging service: ');
        //         console.log(error);
        //     },
        //     () => {
        //         // console.log('log rest service completed!');
        //     }
        // );
    };
    /**
     * Return true if the image size is within range
     * @param file
     */
    /**
     * Return true if the image size is within range
     * @param {?} file
     * @return {?}
     */
    FileUploaderComponent.prototype.checkImageDimensions = /**
     * Return true if the image size is within range
     * @param {?} file
     * @return {?}
     */
    function (file) {
        if (file.naturalHeight < 0 ||
            file.naturalWidth < 0) {
            return false;
        }
        return true;
    };
    /**
     * @return {?}
     */
    FileUploaderComponent.prototype.isValid = /**
     * @return {?}
     */
    function () {
        console.log('isValid', this.images);
        if (this.required) {
            return this.images && this.images.length > 0;
        }
        return true;
    };
    FileUploaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-file-uploader',
                    template: "<div class=\"dropzone\" #dropZone [ngClass]='{\"has-error\": fileControl?.touched && fileControl?.errors?.required}'>\n\n  <div class=\"instruction-zone d-flex align-items-center flex-wrap flex-sm-nowrap flex-column flex-sm-row\">\n    <i class=\"fa fa-cloud-upload fa-4x d-inline-block upload-icon\" aria-hidden=\"true\" #imagePlaceholderRef tabindex=\"0\"></i>\n\n    <input type=\"file\" id=\"fileUploadBrowse-{{id}}\" #browseFileRef ngModel accept=\"image/*,application/pdf\" style=\"display:none;\"\n            tabindex=\"0\" multiple name='fileUploadBrowse-{{id}}' [required]='required && images.length === 0' />\n    <label for=\"fileUploadBrowse-{{id}}\" class=\"file-upload-label d-inline-block ml-3\" #selectFileLabel tabindex=\"0\">\n      <span class='h2 color-body'>Select a file</span>\n      <span class='d-block description'>{{ instructionText }}</span>\n    </label>\n    \n    <!-- TODO - STYLE/ENABLE/TEST THE BELOW FOR MOBILE ONLY! Also need to test on a phone after deploying to OpenShift-->\n    <!-- <span class=\"d-block d-sm-none\">or</span>\n    <input type=\"file\" id=\"fileUploadUseCamera-{{id}}\" #captureFileRef accept=\"image/*,application/pdf\" capture=\"camera\"\n            tabindex=\"0\" multiple />\n    <span class=\"fa fa-camera d-block d-sm-none fa-4x\" aria-hidden=\"true\"></span>\n    <label for=\"fileUploadUseCamera-{{id}}\" class=\"file-upload-label description d-block d-sm-none\" tabindex=\"0\">\n      take a photo\n    </label> -->\n    \n  </div>\n  \n  <div *ngIf=\"fileControl?.touched\" role=\"alert\" class='error-container' aria-live=\"assertive\">\n    <div class=\"text-danger\" *ngIf=\"fileControl?.errors?.required\">\n      {{errorMessages.required}}\n    </div>\n  </div>\n\n  <div class=\"preview-zone\">\n    <div *ngFor=\"let mspImage of images\" class=\"preview-item\">\n      <common-thumbnail [imageObject]=\"mspImage\" (deleteImage)=\"deleteImage($event)\">\n      </common-thumbnail>\n    </div>\n  </div>\n\n  <!-- The sole purpose of this element is to maintain the drop zone to an \n        large size to display on screen.\n      -->\n  <div class=\"preview-zone\">\n    <div class=\"preview-item\">\n    </div>\n  </div>\n\n</div>\n\n<!-- This hidden canvas is used to transform / resize images -->\n<canvas #canvas style=\"display:none;\">\n\n</canvas>\n",
                    viewProviders: [{ provide: ControlContainer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return NgForm; })) }],
                    styles: [".color-body{color:#494949}.description{font-weight:400}.dropzone{border:2px dashed #d3d3d3;margin-bottom:10px;border-radius:8px;padding:2em 4em}.dropzone.has-error{border-color:#a94442}.dropzone .instruction-zone{margin:10px 5px 5px;text-align:left}.dropzone .instruction-zone input{display:none}.dropzone .instruction-zone .file-upload-label{color:#494949}.dropzone .instruction-zone .file-upload-label:hover{text-decoration:none;cursor:pointer}.dropzone .preview-zone{display:flex;flex-wrap:wrap;justify-content:left;align-items:left}.dropzone .preview-zone .preview-item{position:relative;height:120px;text-align:left;margin-left:6rem;margin-bottom:1.5rem}.dropzone .preview-zone .preview-item .icon-upload{opacity:.3;margin:0 auto 15px}.dropzone .preview-zone .preview-item .icon-upload:hover{cursor:pointer;opacity:.6}.dropzone .preview-zone .preview-item .text-upload:hover{cursor:pointer}.dropzone .preview-zone .error-style{color:red;font-weight:700}.dropzone .mobileShow{display:none}@media only screen and (min-device-width:320px) and (max-device-width:480px){.dropzone .mobileShow{display:inline}}"]
                }] }
    ];
    /** @nocollapse */
    FileUploaderComponent.ctorParameters = function () { return [
        { type: NgZone },
        { type: ChangeDetectorRef },
        { type: Router },
        { type: ControlContainer }
    ]; };
    FileUploaderComponent.propDecorators = {
        dropZone: [{ type: ViewChild, args: ['dropZone',] }],
        browseFileRef: [{ type: ViewChild, args: ['browseFileRef',] }],
        imagePlaceholderRef: [{ type: ViewChild, args: ['imagePlaceholderRef',] }],
        selectFileLabelRef: [{ type: ViewChild, args: ['selectFileLabel',] }],
        images: [{ type: Input }],
        imagesChange: [{ type: Output }],
        id: [{ type: Input }],
        showError: [{ type: Input }],
        required: [{ type: Input }],
        instructionText: [{ type: Input }],
        errorMessages: [{ type: Input }],
        canvas: [{ type: ViewChild, args: ['canvas',] }],
        errorDocument: [{ type: Output }]
    };
    return FileUploaderComponent;
}(Base));
export { FileUploaderComponent };
if (false) {
    /** @type {?} */
    FileUploaderComponent.prototype.noIdImage;
    /**
     * @type {?}
     * @private
     */
    FileUploaderComponent.prototype.appConstants;
    /** @type {?} */
    FileUploaderComponent.prototype.dropZone;
    /** @type {?} */
    FileUploaderComponent.prototype.browseFileRef;
    /** @type {?} */
    FileUploaderComponent.prototype.imagePlaceholderRef;
    /** @type {?} */
    FileUploaderComponent.prototype.selectFileLabelRef;
    /** @type {?} */
    FileUploaderComponent.prototype.images;
    /** @type {?} */
    FileUploaderComponent.prototype.imagesChange;
    /** @type {?} */
    FileUploaderComponent.prototype.id;
    /** @type {?} */
    FileUploaderComponent.prototype.showError;
    /** @type {?} */
    FileUploaderComponent.prototype.required;
    /** @type {?} */
    FileUploaderComponent.prototype.instructionText;
    /** @type {?} */
    FileUploaderComponent.prototype.errorMessages;
    /** @type {?} */
    FileUploaderComponent.prototype.canvas;
    /** @type {?} */
    FileUploaderComponent.prototype.errorDocument;
    /**
     * @type {?}
     * @private
     */
    FileUploaderComponent.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    FileUploaderComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    FileUploaderComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    FileUploaderComponent.prototype.controlContainer;
}
