// moh-common-lib/esm5/services/abstract-api-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { catchError, tap } from 'rxjs/operators';
/**
 * Abstract class for HTTP Service
 * @abstract
 */
var /**
 * Abstract class for HTTP Service
 * @abstract
 */
AbstractHttpService = /** @class */ (function () {
    function AbstractHttpService(http) {
        this.http = http;
        this.logHTTPRequestsToConsole = false;
    }
    /**
     * Makes a GET request to the specified URL, using headers and HTTP options specified in their respective methods.
     * @param url Target URL to make the GET request
     */
    /**
     * Makes a GET request to the specified URL, using headers and HTTP options specified in their respective methods.
     * @protected
     * @template T
     * @param {?} url Target URL to make the GET request
     * @param {?=} queryParams
     * @return {?}
     */
    AbstractHttpService.prototype.get = /**
     * Makes a GET request to the specified URL, using headers and HTTP options specified in their respective methods.
     * @protected
     * @template T
     * @param {?} url Target URL to make the GET request
     * @param {?=} queryParams
     * @return {?}
     */
    function (url, queryParams) {
        /**
         * The HTTP request observer with always on error handling
         * @type {?}
         */
        var httpOpts = this.httpOptions;
        httpOpts.params = queryParams ? queryParams : undefined;
        /** @type {?} */
        var observable = this.http.get(url, httpOpts);
        return this.setupRequest(observable);
    };
    /**
     * @protected
     * @template T
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    AbstractHttpService.prototype.post = /**
     * @protected
     * @template T
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        if (this.logHTTPRequestsToConsole) {
            console.log('Post Request: ', body);
        }
        /** @type {?} */
        var observable = this.http.post(url, body, this.httpOptions);
        return this.setupRequest(observable);
    };
    /**
     * @protected
     * @template T
     * @param {?} observable
     * @return {?}
     */
    AbstractHttpService.prototype.setupRequest = /**
     * @protected
     * @template T
     * @param {?} observable
     * @return {?}
     */
    function (observable) {
        // All failed requests should trigger the abstract method handleError
        observable = observable.pipe(catchError(this.handleError.bind(this)));
        // Optionally add console logging
        if (this.logHTTPRequestsToConsole) {
            observable = observable.pipe(tap((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return console.log('HTTP Success: ', data); }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return console.log('HTTP Error: ', error); })));
        }
        return observable;
    };
    Object.defineProperty(AbstractHttpService.prototype, "httpOptions", {
        /** The HttpOptions object that Angular takes for GET and POST requests. Used in every HTTP request from this service. */
        get: /**
         * The HttpOptions object that Angular takes for GET and POST requests. Used in every HTTP request from this service.
         * @protected
         * @return {?}
         */
        function () {
            return {
                headers: this._headers
            };
        },
        enumerable: true,
        configurable: true
    });
    return AbstractHttpService;
}());
/**
 * Abstract class for HTTP Service
 * @abstract
 */
export { AbstractHttpService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype.logHTTPRequestsToConsole;
    /**
     * The headers to send along with every GET and POST.
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype._headers;
    /**
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype.http;
    /**
     * Handles all failed requests that throw either a server error (400/500) or a client error (e.g. lost internet).
     * @abstract
     * @protected
     * @param {?} error
     * @return {?}
     */
    AbstractHttpService.prototype.handleError = function (error) { };
}
