// moh-common-lib/esm5/services/logger.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { AbstractHttpService } from './abstract-api-service';
import { throwError } from 'rxjs';
import * as moment_ from 'moment';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/** @type {?} */
var moment = moment_;
var CommonLogger = /** @class */ (function (_super) {
    tslib_1.__extends(CommonLogger, _super);
    function CommonLogger(http) {
        var _this = _super.call(this, http) || this;
        _this.http = http;
        /**
         * The HTTP Headers which go with each request.  These MUST be set if you are
         * using the logger.  Fields include:
         *
         * - program (REQUIRED, the application wide code)
         * - applicationId (REQUIRED, like sessionId)
         * - request_method (REQUIRED, 'POST')
         * - logsource: (REQUIRED, window.location.hostname)
         * - http_x_forwarded_host (REQUIRED, window.location.hostname)
         *
         */
        _this._headers = new HttpHeaders({
            request_method: 'POST',
            logsource: window.location.hostname,
            http_x_forwarded_host: window.location.hostname
        });
        _this.url = null;
        return _this;
    }
    Object.defineProperty(CommonLogger.prototype, "applicationId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._headers.get('applicationId');
        },
        set: /**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            this._headers = this._headers.set('applicationId', id);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CommonLogger.prototype, "programName", {
        get: /**
         * @return {?}
         */
        function () {
            return this._headers.get('name');
        },
        set: /**
         * @param {?} name
         * @return {?}
         */
        function (name) {
            this._headers = this._headers.set('program', name);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} newURL
     * @return {?}
     */
    CommonLogger.prototype.setURL = /**
     * @param {?} newURL
     * @return {?}
     */
    function (newURL) {
        this.url = newURL;
    };
    /**
     * @param {?} message
     * @return {?}
     */
    CommonLogger.prototype.log = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this._log((/** @type {?} */ (message)));
    };
    /**
     * @param {?} errorMessage
     * @return {?}
     */
    CommonLogger.prototype.logError = /**
     * @param {?} errorMessage
     * @return {?}
     */
    function (errorMessage) {
        this._logError((/** @type {?} */ (errorMessage)));
    };
    /**
     * Log a message to Splunk. This is the main way to send logs and
     * automatically includes meta-data. You do **not** need to subscribe to the
     * response, as the service already does that. The input object must have an
     * 'event' property set, everything else is optional.
     *
     * Example:
     * ```
      this.logService.log({
         event: 'submission',
         dateObj: new Date()
      });
      ```
     * @param message A JavaScript object, nesting is fine, with `event` property
     * set.
     */
    /**
     * Log a message to Splunk. This is the main way to send logs and
     * automatically includes meta-data. You do **not** need to subscribe to the
     * response, as the service already does that. The input object must have an
     * 'event' property set, everything else is optional.
     *
     * Example:
     * ```
     * this.logService.log({
     * event: 'submission',
     * dateObj: new Date()
     * });
     * ```
     * @protected
     * @param {?} message A JavaScript object, nesting is fine, with `event` property
     * set.
     * @return {?}
     */
    CommonLogger.prototype._log = /**
     * Log a message to Splunk. This is the main way to send logs and
     * automatically includes meta-data. You do **not** need to subscribe to the
     * response, as the service already does that. The input object must have an
     * 'event' property set, everything else is optional.
     *
     * Example:
     * ```
     * this.logService.log({
     * event: 'submission',
     * dateObj: new Date()
     * });
     * ```
     * @protected
     * @param {?} message A JavaScript object, nesting is fine, with `event` property
     * set.
     * @return {?}
     */
    function (message) {
        this.setSeverity(SeverityLevels.INFO);
        return this._sendLog(message);
    };
    /**
     * @protected
     * @param {?} errorMessage
     * @return {?}
     */
    CommonLogger.prototype._logError = /**
     * @protected
     * @param {?} errorMessage
     * @return {?}
     */
    function (errorMessage) {
        this.setSeverity(SeverityLevels.ERROR);
        return this._sendLog(errorMessage);
    };
    /**
     * Log HTTP errors, e.g. when losing network connectivity or receiving an
     * error response code.
     */
    /**
     * Log HTTP errors, e.g. when losing network connectivity or receiving an
     * error response code.
     * @param {?} error
     * @return {?}
     */
    CommonLogger.prototype.logHttpError = /**
     * Log HTTP errors, e.g. when losing network connectivity or receiving an
     * error response code.
     * @param {?} error
     * @return {?}
     */
    function (error) {
        return this._logError({
            event: 'error',
            message: error.message,
            errorName: error.name,
            statusText: error.statusText
        });
    };
    /**
     * Internal method to send logs to Splunk, includes meta-data except that's
     * consistent across all requests, but not specific values like severity
     * level.
     *
     * @param message A JavaScript object or anything that can be toString()'d,
     * like Date
     */
    /**
     * Internal method to send logs to Splunk, includes meta-data except that's
     * consistent across all requests, but not specific values like severity
     * level.
     *
     * @private
     * @param {?} message A JavaScript object or anything that can be toString()'d,
     * like Date
     * @return {?}
     */
    CommonLogger.prototype._sendLog = /**
     * Internal method to send logs to Splunk, includes meta-data except that's
     * consistent across all requests, but not specific values like severity
     * level.
     *
     * @private
     * @param {?} message A JavaScript object or anything that can be toString()'d,
     * like Date
     * @return {?}
     */
    function (message) {
        // Update headers
        this.setTimestamp();
        this.setTags(message.event);
        if (this.url === null) {
            /** @type {?} */
            var msg = 'Unable to send logs as URL as not been set via setURL()';
            console.error(msg);
            return throwError(msg);
        }
        // Configure request
        /** @type {?} */
        var body = { message: message };
        // We call .subscribe() here because we don't care about the response and
        // we want to ensure that we never forget to call subscribe.
        return this.post(this.url, body).subscribe();
    };
    /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    CommonLogger.prototype.handleError = /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    function (error) {
        console.log('logService handleError()', error);
        if (error.error instanceof ErrorEvent) {
            // Client-side / network error occured
            console.error('An error occured: ', error.error.message);
        }
        else {
            // The backend returned an unsuccessful response code
            console.error("Backend returned error code: " + error.status + ".  Error body: " + error.error);
        }
        return throwError(error);
    };
    Object.defineProperty(CommonLogger.prototype, "httpOptions", {
        /**
         * Overwrite the inherited httpOptions so we can set responseType to text.
         * This updates Angular's parsing, and it won't error out due to the server
         * not responding with JSON.
         */
        get: /**
         * Overwrite the inherited httpOptions so we can set responseType to text.
         * This updates Angular's parsing, and it won't error out due to the server
         * not responding with JSON.
         * @protected
         * @return {?}
         */
        function () {
            return {
                headers: this._headers,
                responseType: 'text'
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @return {?}
     */
    CommonLogger.prototype.setTimestamp = /**
     * @private
     * @return {?}
     */
    function () {
        this._headers = this._headers.set('timestamp', moment().toISOString());
    };
    /**
     * @private
     * @param {?} severity
     * @return {?}
     */
    CommonLogger.prototype.setSeverity = /**
     * @private
     * @param {?} severity
     * @return {?}
     */
    function (severity) {
        this._headers = this._headers.set('severity', severity);
    };
    /**
     * The headers are easier to search in splunk, and we aren't using tags, so
     * repurpose it to event type.
     */
    /**
     * The headers are easier to search in splunk, and we aren't using tags, so
     * repurpose it to event type.
     * @private
     * @param {?} message
     * @return {?}
     */
    CommonLogger.prototype.setTags = /**
     * The headers are easier to search in splunk, and we aren't using tags, so
     * repurpose it to event type.
     * @private
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this._headers = this._headers.set('tags', message);
    };
    CommonLogger.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CommonLogger.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ CommonLogger.ngInjectableDef = i0.defineInjectable({ factory: function CommonLogger_Factory() { return new CommonLogger(i0.inject(i1.HttpClient)); }, token: CommonLogger, providedIn: "root" });
    return CommonLogger;
}(AbstractHttpService));
export { CommonLogger };
if (false) {
    /**
     * The HTTP Headers which go with each request.  These MUST be set if you are
     * using the logger.  Fields include:
     *
     * - program (REQUIRED, the application wide code)
     * - applicationId (REQUIRED, like sessionId)
     * - request_method (REQUIRED, 'POST')
     * - logsource: (REQUIRED, window.location.hostname)
     * - http_x_forwarded_host (REQUIRED, window.location.hostname)
     *
     * @type {?}
     * @protected
     */
    CommonLogger.prototype._headers;
    /**
     * @type {?}
     * @private
     */
    CommonLogger.prototype.url;
    /**
     * @type {?}
     * @protected
     */
    CommonLogger.prototype.http;
}
/** @enum {string} */
var SeverityLevels = {
    INFO: 'info',
    ERROR: 'error',
};
/**
 * @record
 */
export function CommonLogMessage() { }
if (false) {
    /**
     * The type of event being logged.
     * @type {?}
     */
    CommonLogMessage.prototype.event;
    /* Skipping unhandled member: [key: string]: any;*/
}
