// moh-common-lib/esm2015/services/abstract-api-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { catchError, tap } from 'rxjs/operators';
/**
 * Abstract class for HTTP Service
 * @abstract
 */
export class AbstractHttpService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.logHTTPRequestsToConsole = false;
    }
    /**
     * Makes a GET request to the specified URL, using headers and HTTP options specified in their respective methods.
     * @protected
     * @template T
     * @param {?} url Target URL to make the GET request
     * @param {?=} queryParams
     * @return {?}
     */
    get(url, queryParams) {
        /**
         * The HTTP request observer with always on error handling
         * @type {?}
         */
        const httpOpts = this.httpOptions;
        httpOpts.params = queryParams ? queryParams : undefined;
        /** @type {?} */
        const observable = this.http.get(url, httpOpts);
        return this.setupRequest(observable);
    }
    /**
     * @protected
     * @template T
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    post(url, body) {
        if (this.logHTTPRequestsToConsole) {
            console.log('Post Request: ', body);
        }
        /** @type {?} */
        const observable = this.http.post(url, body, this.httpOptions);
        return this.setupRequest(observable);
    }
    /**
     * @protected
     * @template T
     * @param {?} observable
     * @return {?}
     */
    setupRequest(observable) {
        // All failed requests should trigger the abstract method handleError
        observable = observable.pipe(catchError(this.handleError.bind(this)));
        // Optionally add console logging
        if (this.logHTTPRequestsToConsole) {
            observable = observable.pipe(tap((/**
             * @param {?} data
             * @return {?}
             */
            data => console.log('HTTP Success: ', data)), (/**
             * @param {?} error
             * @return {?}
             */
            error => console.log('HTTP Error: ', error))));
        }
        return observable;
    }
    /**
     * The HttpOptions object that Angular takes for GET and POST requests. Used in every HTTP request from this service.
     * @protected
     * @return {?}
     */
    get httpOptions() {
        return {
            headers: this._headers
        };
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype.logHTTPRequestsToConsole;
    /**
     * The headers to send along with every GET and POST.
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype._headers;
    /**
     * @type {?}
     * @protected
     */
    AbstractHttpService.prototype.http;
    /**
     * Handles all failed requests that throw either a server error (400/500) or a client error (e.g. lost internet).
     * @abstract
     * @protected
     * @param {?} error
     * @return {?}
     */
    AbstractHttpService.prototype.handleError = function (error) { };
}
