// moh-common-lib/esm2015/services/geocoder.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { AbstractHttpService } from './abstract-api-service';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { throwError } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/**
 * @record
 */
export function GeoAddressResult() { }
if (false) {
    /**
     * String from the API that includes street, city, province, and country.
     * @type {?}
     */
    GeoAddressResult.prototype.fullAddress;
    /** @type {?} */
    GeoAddressResult.prototype.city;
    /** @type {?} */
    GeoAddressResult.prototype.street;
    /** @type {?} */
    GeoAddressResult.prototype.country;
    /** @type {?} */
    GeoAddressResult.prototype.province;
}
export class GeocoderService extends AbstractHttpService {
    /**
     * @param {?} http
     */
    constructor(http) {
        super(http);
        this.http = http;
        this._headers = new HttpHeaders();
        this.BASE_URL = 'https://geocoder.api.gov.bc.ca';
        this.ADDRESS_URL = `${this.BASE_URL}/addresses.json?`;
        /**
         * Defaults for service
         */
        this.defaultCountry = 'Canada';
        this.defaultProv = 'British Columbia';
    }
    // https://geocoder.api.gov.bc.ca/addresses.json?minScore=50&maxResults=5&echo=false&brief=true&autoComplete=true&addressString=784+Hock
    /**
     * @param {?} address
     * @return {?}
     */
    lookup(address) {
        /** @type {?} */
        const params = new HttpParams()
            .set('minScore', '50')
            .set('maxResults', '10')
            .set('echo', 'false')
            .set('brief', 'false') // API splits address string up into sub-attributes, like city  / street name
            .set('autoComplete', 'true')
            .set('matchPrecisionNot', 'LOCALITY,STREET,BLOCK,INTERSECTION')
            .set('addressString', address);
        return this.get(this.ADDRESS_URL, params).pipe(map(this.processResponse));
    }
    /**
     * Formats the response from ADDRESS_URL, trimming irrelevant fields.
     *
     * This works for other requests for the same API too, however it may error
     * out on some items if matchPrecisionNot is not set.
     *
     * @private
     * @param {?} obj The response from ADDRESS_URL
     * @return {?}
     */
    processResponse(obj) {
        return obj.features.map((/**
         * @param {?} feature
         * @return {?}
         */
        feature => {
            /** @type {?} */
            const props = feature.properties;
            /** @type {?} */
            const city = props.localityName;
            // We get street just by trimming everything before city, more
            // stable than looking for commas, etc.
            /** @type {?} */
            const cityIndex = props.fullAddress.indexOf(`, ${city}`);
            /** @type {?} */
            const street = props.fullAddress.slice(0, cityIndex);
            return {
                fullAddress: props.fullAddress,
                city,
                street,
                country: this.defaultCountry,
                // Default to Canada
                province: this.defaultProv // Default to BC
            };
        }));
    }
    /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    handleError(error) {
        console.error('GeoCoder network error', { error });
        return throwError('Geocoder error');
    }
}
GeocoderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
GeocoderService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ GeocoderService.ngInjectableDef = i0.defineInjectable({ factory: function GeocoderService_Factory() { return new GeocoderService(i0.inject(i1.HttpClient)); }, token: GeocoderService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    GeocoderService.prototype._headers;
    /**
     * @type {?}
     * @private
     */
    GeocoderService.prototype.BASE_URL;
    /**
     * @type {?}
     * @private
     */
    GeocoderService.prototype.ADDRESS_URL;
    /**
     * Defaults for service
     * @type {?}
     */
    GeocoderService.prototype.defaultCountry;
    /** @type {?} */
    GeocoderService.prototype.defaultProv;
    /**
     * @type {?}
     * @protected
     */
    GeocoderService.prototype.http;
}
