// moh-common-lib/esm2015/models/abstract-form.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
/**
 *
 * @abstract
 */
export class AbstractForm {
    /**
     * Constructor
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        /**
         * Show or hide the loading spinner as required, should be passed to form action bar.
         */
        this.loading = false;
        this.router = router;
    }
    /**
     * Determines if the Continue button is disabled on the form action bar
     * @return {?}
     */
    canContinue() {
        // Returns true if form is valid
        return this.form.valid;
    }
    /**
     * Navigates to a route then automatically scrolls to the top of the page.
     * @protected
     * @param {?} url
     * @return {?}
     */
    navigate(url) {
        this.router.navigate([url])
            .then((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            window.scrollTo({ top: 0, behavior: 'smooth' });
        }));
    }
    /**
     * Runs the angular 'markAsTouched()' on all form inputs.
     * @protected
     * @return {?}
     */
    markAllInputsTouched() {
        Object.keys(this.form.form.controls).forEach((/**
         * @param {?} x
         * @return {?}
         */
        x => {
            this.form.form.get(x).markAsTouched();
        }));
    }
}
AbstractForm.propDecorators = {
    form: [{ type: ViewChild, args: ['formRef',] }]
};
if (false) {
    /**
     * Access to the form elements for validation
     * @type {?}
     */
    AbstractForm.prototype.form;
    /**
     * Disables all inputs (todo: not implemented)
     * @type {?}
     */
    AbstractForm.prototype.disabled;
    /**
     * Show or hide the loading spinner as required, should be passed to form action bar.
     * @type {?}
     */
    AbstractForm.prototype.loading;
    /**
     * @type {?}
     * @protected
     */
    AbstractForm.prototype.router;
    /**
     * What happens when the user clicks the continue button. Generally navigating to another page.
     * @abstract
     * @return {?}
     */
    AbstractForm.prototype.continue = function () { };
}
