// moh-common-lib/esm2015/models/address.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Address class, each project can extend this address class
 */
export class Address {
    /**
     * Overwrite the native JavaScript toString method to determine how the
     * object should be printed, instead of [object Object].  This provides a
     * standard way to print out an address. If you need something specific you
     * should access the properties directly. We omit Province/Country because of
     * PharmaCare's BC focus.
     * @return {?}
     */
    toString() {
        return `${this.street}, ${this.city}`;
    }
    /**
     * Address must have all fields filled out to be considered
     * complete
     * @return {?}
     */
    isComplete() {
        // All fields have data - not empty
        return !!(this.street && this.city && this.country &&
            this.province && this.postal);
    }
    /* Copy function */
    /**
     * @param {?} object
     * @return {?}
     */
    copy(object) {
        this.street = object.street;
        this.city = object.city;
        this.country = object.country;
        this.postal = object.postal;
        this.province = object.province;
    }
}
if (false) {
    /** @type {?} */
    Address.prototype.street;
    /** @type {?} */
    Address.prototype.postal;
    /** @type {?} */
    Address.prototype.country;
    /** @type {?} */
    Address.prototype.province;
    /** @type {?} */
    Address.prototype.city;
}
