// moh-common-lib/esm2015/models/person.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as moment_ from 'moment';
import { Base } from './base';
/** @type {?} */
const moment = moment_;
/**
 * Person, each project can extend this person class
 * NPM package dependencies:
 *  a) moment
 */
export class Person extends Base {
    constructor() {
        super(...arguments);
        /**
         * Format to display birthdate
         */
        this.dobFormat = 'YYYY/MM/DD';
        // Initialize dob to nulls
        this.dateOfBirth = { year: null, month: null, day: null };
    }
    /**
     * Returns DoB in YYYYMMDD format, used by API.
     * @return {?}
     */
    get dateOfBirthShort() {
        return this.isDobEmpty() ? null :
            moment({
                year: this.dateOfBirth.year,
                month: this.dateOfBirth.month - 1,
                day: this.dateOfBirth.day
            }).format('YYYYMMDD');
    }
    /**
     * Returns DoB in dobFormat (default: YYYY/MM/DD), for display purposes
     * @return {?}
     */
    get formatDateOfBirth() {
        return this.isDobEmpty() ? null :
            moment({
                year: this.dateOfBirth.year,
                month: this.dateOfBirth.month - 1,
                day: this.dateOfBirth.day
            }).format(this.dobFormat);
    }
    /**
     * Indicates whether or not the date of birth is empty
     * @return {?}
     */
    isDobEmpty() {
        return Object.keys(this.dateOfBirth)
            .map((/**
         * @param {?} key
         * @return {?}
         */
        key => this.dateOfBirth[key]))
            .filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x)) // Filter out null/undefined
            .length !== 3;
    }
    /**
     * Concatenates the first and last name together
     * @return {?}
     */
    get name() {
        /** @type {?} */
        let _name = null;
        if (this.firstName) {
            _name = this.firstName;
        }
        if (this.lastName) {
            _name = _name ? _name.concat(' ' + this.lastName) : this.lastName;
        }
        return _name;
    }
    /**
     * Sets the full name for the person (first, middle and last names)
     * NOTE: Just for development with dummy data and unit tests
     * @param {?} fullName
     * @return {?}
     */
    set name(fullName) {
        /** @type {?} */
        const names = fullName.split(' ');
        this.firstName = names[0];
        if (names.length === 2) {
            this.lastName = names[1];
        }
        else if (names.length === 3) {
            this.middleName = names[1];
            this.lastName = names[2];
        }
    }
    /**
     * Calculates the age from date of birth
     * @return {?}
     */
    getAge() {
        /** @type {?} */
        const dobDt = new Date(this.dateOfBirth.year, this.dateOfBirth.month, this.dateOfBirth.day);
        return Math.ceil(moment().diff(dobDt, 'year', true));
    }
    /* Copy function */
    /**
     * @param {?} object
     * @return {?}
     */
    copy(object) {
        this.firstName = object.firstName;
        this.middleName = object.middleName;
        this.lastName = object.lastName;
        this.dateOfBirth.month = object.dateOfBirth.month;
        this.dateOfBirth.day = object.dateOfBirth.day;
        this.dateOfBirth.year = object.dateOfBirth.year;
    }
}
if (false) {
    /** @type {?} */
    Person.prototype.firstName;
    /** @type {?} */
    Person.prototype.middleName;
    /** @type {?} */
    Person.prototype.lastName;
    /**
     * Format to display birthdate
     * @type {?}
     */
    Person.prototype.dobFormat;
    /** @type {?} */
    Person.prototype.dateOfBirth;
}
