// moh-common-lib/esm5/lib/components/toggle/toggle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Base } from '../../../models/src/base';
var ToggleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ToggleComponent, _super);
    function ToggleComponent() {
        var _this = _super.call(this) || this;
        _this.dataChange = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    ToggleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    ToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-toggle',
                    template: "<fieldset class=\"p-sm-1\">\n   <legend class=\"toggle--legend\">{{label}}</legend>\n   <div class=\"custom-radio custom-control custom-control-inline\">\n      <input class=\"custom-control-input \"\n         type=\"radio\"\n         id=\"no_{{objectId}}\"\n         name=\"selection_{{objectId}}\"\n         value=\"No\"\n         [checked]=\"data === false\"\n         (click)=\"dataChange.emit(false)\">\n      <label class=\"custom-control-label\" for=\"no_{{objectId}}\">No</label>\n   </div>\n   <div class=\"custom-radio custom-control custom-control-inline\">\n      <input class=\"custom-control-input\"\n         type=\"radio\"\n         id=\"yes_{{objectId}}\"\n         name=\"selection_{{objectId}}\"\n         value=\"Yes\"\n         [checked]=\"data === true\"\n         (click)=\"dataChange.emit(true)\">\n      <label class=\"custom-control-label\" for=\"yes_{{objectId}}\">Yes</label>\n   </div>\n</fieldset>\n\n",
                    styles: [".toggle--legend{font-size:inherit;font-weight:700}.custom-control{padding-right:1.5rem;padding-left:0}.custom-control-label{cursor:pointer}.custom-radio .custom-control-label::before{border:1.5px solid #000}.custom-control-label::before{top:0;width:1.3rem;height:1.3rem;background-color:#fcfcfc;left:1.75rem}.custom-control-label::after{top:0;width:1.3rem;height:1.3rem;left:1.75rem}.custom-radio .custom-control-input:checked~.custom-control-label::before{background-color:#fcfcfc;border-color:#036}.custom-control-input:checked~.custom-control-label::before{color:#fcfcfc;background-color:#fcfcfc;border-color:#036}.custom-radio .custom-control-input:checked~.custom-control-label::after{background-image:url(\"data:image/svg+xml;charset=utf8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='-3.75 -3.75 8.25 8.25'%3E%3Ccircle r='3.5' fill='%23000099'/%3E%3C/svg%3E\")!important}"]
                }] }
    ];
    /** @nocollapse */
    ToggleComponent.ctorParameters = function () { return []; };
    ToggleComponent.propDecorators = {
        data: [{ type: Input }],
        label: [{ type: Input }],
        dataChange: [{ type: Output }]
    };
    return ToggleComponent;
}(Base));
export { ToggleComponent };
if (false) {
    /** @type {?} */
    ToggleComponent.prototype.data;
    /** @type {?} */
    ToggleComponent.prototype.label;
    /** @type {?} */
    ToggleComponent.prototype.dataChange;
}
