// moh-common-lib/esm2015/lib/components/dropdown/dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class DropdownComponent {
    constructor() {
        this.items = [];
        this.label = 'Default label';
        this.addTag = false;
        this.addTagText = 'Add';
        this._onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => { });
        this._onTouched = (/**
         * @return {?}
         */
        () => { });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.model = value;
        // TODO - is markForCheck() necessary?
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
}
DropdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'common-dropdown',
                template: "<ng-select name='{{label}}'\n           labelForId='{{label}}'\n           [ngModel]='model'\n           (ngModelChange)='model=$event;'\n           [items]='items'\n           placeholder='{{ placeholder }}'\n           autoCorrect='{{ autocorrect }}'\n           addTag={{addTag}}\n           addTagText='Add question'\n           [required]='required'\n           (change)='_onChange($event)'\n           (blur)='_onTouched()'\n           (close)='_onTouched()'></ng-select>",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => DropdownComponent))
                    }
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
DropdownComponent.ctorParameters = () => [];
DropdownComponent.propDecorators = {
    items: [{ type: Input }],
    label: [{ type: Input }],
    placeholder: [{ type: Input }],
    autocorrect: [{ type: Input }],
    addTag: [{ type: Input }],
    addTagText: [{ type: Input }],
    required: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DropdownComponent.prototype.model;
    /** @type {?} */
    DropdownComponent.prototype.items;
    /** @type {?} */
    DropdownComponent.prototype.label;
    /** @type {?} */
    DropdownComponent.prototype.placeholder;
    /** @type {?} */
    DropdownComponent.prototype.autocorrect;
    /** @type {?} */
    DropdownComponent.prototype.addTag;
    /** @type {?} */
    DropdownComponent.prototype.addTagText;
    /** @type {?} */
    DropdownComponent.prototype.required;
    /** @type {?} */
    DropdownComponent.prototype._onChange;
    /** @type {?} */
    DropdownComponent.prototype._onTouched;
}
