// moh-common-lib/esm5/models/abstract-form.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
/**
 *
 * @abstract
 */
var AbstractForm = /** @class */ (function () {
    /**
     * Constructor
     */
    function AbstractForm(router) {
        this.router = router;
        /**
         * Show or hide the loading spinner as required, should be passed to form action bar.
         */
        this.loading = false;
        this.router = router;
    }
    /**
     * Determines if the Continue button is disabled on the form action bar
     */
    /**
     * Determines if the Continue button is disabled on the form action bar
     * @return {?}
     */
    AbstractForm.prototype.canContinue = /**
     * Determines if the Continue button is disabled on the form action bar
     * @return {?}
     */
    function () {
        // Returns true if form is valid
        return this.form.valid;
    };
    /** Navigates to a route then automatically scrolls to the top of the page. */
    /**
     * Navigates to a route then automatically scrolls to the top of the page.
     * @protected
     * @param {?} url
     * @return {?}
     */
    AbstractForm.prototype.navigate = /**
     * Navigates to a route then automatically scrolls to the top of the page.
     * @protected
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.router.navigate([url])
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            window.scrollTo({ top: 0, behavior: 'smooth' });
        }));
    };
    /** Runs the angular 'markAsTouched()' on all form inputs. */
    /**
     * Runs the angular 'markAsTouched()' on all form inputs.
     * @protected
     * @return {?}
     */
    AbstractForm.prototype.markAllInputsTouched = /**
     * Runs the angular 'markAsTouched()' on all form inputs.
     * @protected
     * @return {?}
     */
    function () {
        var _this = this;
        Object.keys(this.form.form.controls).forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            _this.form.form.get(x).markAsTouched();
        }));
    };
    AbstractForm.propDecorators = {
        form: [{ type: ViewChild, args: ['formRef',] }]
    };
    return AbstractForm;
}());
export { AbstractForm };
if (false) {
    /**
     * Access to the form elements for validation
     * @type {?}
     */
    AbstractForm.prototype.form;
    /**
     * Disables all inputs (todo: not implemented)
     * @type {?}
     */
    AbstractForm.prototype.disabled;
    /**
     * Show or hide the loading spinner as required, should be passed to form action bar.
     * @type {?}
     */
    AbstractForm.prototype.loading;
    /**
     * @type {?}
     * @protected
     */
    AbstractForm.prototype.router;
    /**
     * What happens when the user clicks the continue button. Generally navigating to another page.
     * @abstract
     * @return {?}
     */
    AbstractForm.prototype.continue = function () { };
}
