// moh-common-lib/esm5/models/person.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import * as moment_ from 'moment';
import { Base } from './base';
/** @type {?} */
var moment = moment_;
/**
 * Person, each project can extend this person class
 * NPM package dependencies:
 *  a) moment
 */
var /**
 * Person, each project can extend this person class
 * NPM package dependencies:
 *  a) moment
 */
Person = /** @class */ (function (_super) {
    tslib_1.__extends(Person, _super);
    function Person() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Format to display birthdate
         */
        _this.dobFormat = 'YYYY/MM/DD';
        // Initialize dob to nulls
        _this.dateOfBirth = { year: null, month: null, day: null };
        return _this;
    }
    Object.defineProperty(Person.prototype, "dateOfBirthShort", {
        /** Returns DoB in YYYYMMDD format, used by API. */
        get: /**
         * Returns DoB in YYYYMMDD format, used by API.
         * @return {?}
         */
        function () {
            return this.isDobEmpty() ? null :
                moment({
                    year: this.dateOfBirth.year,
                    month: this.dateOfBirth.month - 1,
                    day: this.dateOfBirth.day
                }).format('YYYYMMDD');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Person.prototype, "formatDateOfBirth", {
        /** Returns DoB in dobFormat (default: YYYY/MM/DD), for display purposes */
        get: /**
         * Returns DoB in dobFormat (default: YYYY/MM/DD), for display purposes
         * @return {?}
         */
        function () {
            return this.isDobEmpty() ? null :
                moment({
                    year: this.dateOfBirth.year,
                    month: this.dateOfBirth.month - 1,
                    day: this.dateOfBirth.day
                }).format(this.dobFormat);
        },
        enumerable: true,
        configurable: true
    });
    /** Indicates whether or not the date of birth is empty */
    /**
     * Indicates whether or not the date of birth is empty
     * @return {?}
     */
    Person.prototype.isDobEmpty = /**
     * Indicates whether or not the date of birth is empty
     * @return {?}
     */
    function () {
        var _this = this;
        return Object.keys(this.dateOfBirth)
            .map((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return _this.dateOfBirth[key]; }))
            .filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x; })) // Filter out null/undefined
            .length !== 3;
    };
    Object.defineProperty(Person.prototype, "name", {
        /** Concatenates the first and last name together */
        get: /**
         * Concatenates the first and last name together
         * @return {?}
         */
        function () {
            /** @type {?} */
            var _name = null;
            if (this.firstName) {
                _name = this.firstName;
            }
            if (this.lastName) {
                _name = _name ? _name.concat(' ' + this.lastName) : this.lastName;
            }
            return _name;
        },
        /**
         * Sets the full name for the person (first, middle and last names)
         * NOTE: Just for development with dummy data and unit tests
         */
        set: /**
         * Sets the full name for the person (first, middle and last names)
         * NOTE: Just for development with dummy data and unit tests
         * @param {?} fullName
         * @return {?}
         */
        function (fullName) {
            /** @type {?} */
            var names = fullName.split(' ');
            this.firstName = names[0];
            if (names.length === 2) {
                this.lastName = names[1];
            }
            else if (names.length === 3) {
                this.middleName = names[1];
                this.lastName = names[2];
            }
        },
        enumerable: true,
        configurable: true
    });
    /** Calculates the age from date of birth */
    /**
     * Calculates the age from date of birth
     * @return {?}
     */
    Person.prototype.getAge = /**
     * Calculates the age from date of birth
     * @return {?}
     */
    function () {
        /** @type {?} */
        var dobDt = new Date(this.dateOfBirth.year, this.dateOfBirth.month, this.dateOfBirth.day);
        return Math.ceil(moment().diff(dobDt, 'year', true));
    };
    /* Copy function */
    /* Copy function */
    /**
     * @param {?} object
     * @return {?}
     */
    Person.prototype.copy = /* Copy function */
    /**
     * @param {?} object
     * @return {?}
     */
    function (object) {
        this.firstName = object.firstName;
        this.middleName = object.middleName;
        this.lastName = object.lastName;
        this.dateOfBirth.month = object.dateOfBirth.month;
        this.dateOfBirth.day = object.dateOfBirth.day;
        this.dateOfBirth.year = object.dateOfBirth.year;
    };
    return Person;
}(Base));
/**
 * Person, each project can extend this person class
 * NPM package dependencies:
 *  a) moment
 */
export { Person };
if (false) {
    /** @type {?} */
    Person.prototype.firstName;
    /** @type {?} */
    Person.prototype.middleName;
    /** @type {?} */
    Person.prototype.lastName;
    /**
     * Format to display birthdate
     * @type {?}
     */
    Person.prototype.dobFormat;
    /** @type {?} */
    Person.prototype.dateOfBirth;
}
