// moh-common-lib/esm5/lib/components/wizard-progress-bar/wizard-progress-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef, ViewChildren, QueryList, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';
import { filter, map } from 'rxjs/operators';
/**
 * NPM Dependencies:
 *  a) rxjs
 *  b) ngx-bootstrap
 */
var WizardProgressBarComponent = /** @class */ (function () {
    function WizardProgressBarComponent(router, cd) {
        this.router = router;
        this.cd = cd;
        this.progressSteps = [];
    }
    /**
     * @return {?}
     */
    WizardProgressBarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Update the progress bar view on route change and _only_ route chaange.
        // Skip most of Angular's ChangeDetection in favour of manually optimizing.
        this.routerEvents$ = this.router.events.pipe(filter((/**
         * @param {?} ev
         * @return {?}
         */
        function (ev) { return ev instanceof NavigationEnd; })), map((/**
         * @param {?} ev
         * @return {?}
         */
        function (ev) { return ev.url; }))).subscribe((/**
         * @param {?} url
         * @return {?}
         */
        function (url) {
            _this.activeIndex = _this.getActiveIndex(url);
            _this.cd.detectChanges();
            _this.scrollStepIntoView();
        }));
        // Must schedule first run manually, or bar won't be set.
        this.activeIndex = this.getActiveIndex(this.router.url);
    };
    /**
     * @return {?}
     */
    WizardProgressBarComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.cd.detach();
        this.routerEvents$.unsubscribe();
    };
    /**
     * @return {?}
     */
    WizardProgressBarComponent.prototype.calculateProgressPercentage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var denominator = this.progressSteps.length;
        /** @type {?} */
        var numerator = this.activeIndex + 1;
        if (denominator === 0 || numerator > denominator) {
            return 100;
        }
        // Because we've switched from space-evenly to space-around (for IE), we
        // have to handle the half-space that space-around adds to the start/end of
        // the container
        /** @type {?} */
        var halfSpace = 1 / (denominator * 2);
        return Math.round(((numerator / denominator) - halfSpace) * 100);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    WizardProgressBarComponent.prototype.getActiveIndex = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.progressSteps.findIndex((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return url.includes(x.route); }));
    };
    /**
    * Primarily for mobile, this horizontally scrolls the step into view.
    *
    * Note - be very careful with any changes to this function because it steps
    * outside of Angular to call native browser functions.
    */
    /**
     * Primarily for mobile, this horizontally scrolls the step into view.
     *
     * Note - be very careful with any changes to this function because it steps
     * outside of Angular to call native browser functions.
     * @private
     * @return {?}
     */
    WizardProgressBarComponent.prototype.scrollStepIntoView = /**
     * Primarily for mobile, this horizontally scrolls the step into view.
     *
     * Note - be very careful with any changes to this function because it steps
     * outside of Angular to call native browser functions.
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var target = this.steps.toArray()[this.activeIndex];
        /** @type {?} */
        var container = document.getElementsByClassName('horizontal-scroll');
        if (container.length === 1) {
            // Since we're already breaking out of Angular, we try and be safe by using a try/catch.
            // Otherwise an error here could halt execution,
            try {
                container[0].scrollLeft = Math.abs(target.nativeElement.offsetLeft - (window.outerWidth / 2));
            }
            catch (error) { }
        }
    };
    WizardProgressBarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-wizard-progress-bar',
                    template: "<progressbar\n  [value]=\"calculateProgressPercentage()\"\n  [max]=\"100\"\n  [animate]=\"true\">\n</progressbar>\n\n<div class=\"step-container\" #stepContainer>\n\n  <a #steps *ngFor=\"let step of progressSteps; let i = index;\" [routerLink]=\"step.route\">\n\n    <div class=\"step\" [ngClass]=\"{active: i === activeIndex}\">\n      <span class=\"step-text\">{{step.title}}</span>\n    </div>\n\n  </a>\n\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{flex:1;padding:0 2em;min-height:2em;min-width:650px}.step-container{display:flex;justify-content:space-around}progressbar{background-color:#adb5bd;height:.5rem}.step{position:relative;-webkit-transform:translateX(-.5em);transform:translateX(-.5em);margin-top:.25rem}.step:before{content:\" \";position:absolute;width:1em;height:1em;border-radius:100%;background:#fff;border:3px solid #036;right:0;left:0;margin:0 auto;bottom:100%}.step:not(.active) .step-text{opacity:.8}.step:not(.active):before{background:#ced4da}.step .step-text{position:absolute;-webkit-transform:translateX(-33%);transform:translateX(-33%);white-space:nowrap;font-size:small}.progress-background{width:100%;background-color:#00f}"]
                }] }
    ];
    /** @nocollapse */
    WizardProgressBarComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ChangeDetectorRef }
    ]; };
    WizardProgressBarComponent.propDecorators = {
        progressSteps: [{ type: Input }],
        stepContainer: [{ type: ViewChild, args: ['stepContainer',] }],
        steps: [{ type: ViewChildren, args: ['steps',] }]
    };
    return WizardProgressBarComponent;
}());
export { WizardProgressBarComponent };
if (false) {
    /** @type {?} */
    WizardProgressBarComponent.prototype.progressSteps;
    /** @type {?} */
    WizardProgressBarComponent.prototype.stepContainer;
    /** @type {?} */
    WizardProgressBarComponent.prototype.steps;
    /** @type {?} */
    WizardProgressBarComponent.prototype.activeIndex;
    /**
     * @type {?}
     * @private
     */
    WizardProgressBarComponent.prototype.routerEvents$;
    /**
     * @type {?}
     * @private
     */
    WizardProgressBarComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    WizardProgressBarComponent.prototype.cd;
}
