// moh-common-lib/esm5/lib/components/page-framework/page-framework.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * The "Page Framework" is a template to be used on FPCare pages to ensure
 * consistent layout.  It applies to most pages, but should NOT be used on
 * Dashboards, or full-width table components.
 *
 * \@example
 * <common-page-framework>
 *        <div>This will go in the middle column</div>
 *        <p>So will this</p>
 *        <div aside> This will go in the side column, or tips.</div>
 * </common-page-framework>
 *
 * @export
 */
var PageFrameworkComponent = /** @class */ (function () {
    function PageFrameworkComponent() {
        this.layout = 'default';
    }
    /**
     * @return {?}
     */
    PageFrameworkComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    PageFrameworkComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-page-framework',
                    template: "<div class=\"row\" [ngSwitch]=\"layout\">\n\n  <ng-container *ngSwitchCase=\"'default'\">\n    <div class=\"col-md-8\">\n      <div class=\"px-lg-4 px-md-3 py-3\">\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      </div>\n    </div>\n    <div class=\"col-md-4 aside-col\">\n      <div class='pr-lg-5 pr-md-4 py-2'>\n        <ng-container *ngTemplateOutlet=\"aside\"></ng-container>\n      </div>\n    </div>\n  </ng-container>\n\n\n  <ng-container *ngSwitchCase=\"'double'\">\n    <div class=\"col-md-6\">\n      <div class=\"px-lg-5 px-md-3 py-3\">\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      </div>\n    </div>\n    <div class=\"col-md-6\">\n      <div class='px-lg-5 px-md-3 py-3'>\n        <ng-container *ngTemplateOutlet=\"aside\"></ng-container>\n      </div>\n    </div>\n  </ng-container>\n\n    <ng-container *ngSwitchCase=\"'single'\">\n      <div class=\"col-md-8 offset-lg-2\">\n        <div class=\"px-lg-5 px-md-3 py-3\">\n          <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n        </div>\n      </div>\n    </ng-container>\n\n    <ng-container *ngSwitchCase=\"'blank'\">\n      <div class=\"col-sm-12\">\n        <div class=\"px-lg-5 px-md-3 py-3\">\n          <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n        </div>\n      </div>\n    </ng-container>\n</div>\n<ng-container *ngTemplateOutlet=\"submit\"></ng-container>\n\n\n<!-- We use ng-template here to get around a bug with having multiple ng-contents in one template. By default, if there are duplicate ng-contents in a template Angular will select the very first one - even if latter ones are 'removed' by ngSwitch or ngIf.-->\n<ng-template #content>\n  <ng-content></ng-content>\n</ng-template>\n\n<ng-template #aside>\n    <ng-content select='aside'></ng-content>\n</ng-template>\n\n<ng-template #submit>\n  <ng-content select='common-form-action-bar'></ng-content>\n  <ng-content select='common-form-submit-bar'></ng-content>\n</ng-template>",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["common-page-framework{background:#fcfcfc;display:block}.h5,h5{font-size:1rem;font-weight:700;margin-bottom:0}p+.h5,p+h5{margin-top:1.5rem}.aside-col aside{background:#f2f2f2;padding:1em;border-radius:5px}@media (min-width:768px){.aside-col aside{margin-top:1rem}}"]
                }] }
    ];
    PageFrameworkComponent.propDecorators = {
        layout: [{ type: Input }]
    };
    return PageFrameworkComponent;
}());
export { PageFrameworkComponent };
if (false) {
    /** @type {?} */
    PageFrameworkComponent.prototype.layout;
}
