// moh-common-lib/esm2015/models/src/container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function WizardProgressItem() { }
if (false) {
    /** @type {?} */
    WizardProgressItem.prototype.title;
    /** @type {?} */
    WizardProgressItem.prototype.route;
}
/**
 * Base functionality for container that is used to display bread crumbs
 */
export class Container {
    /**
     * Converts a lower case string of a route in a user readable title.  e.g.
     * "personal-info" -> "Personal Info"
     * @param {?} routePath
     * @return {?}
     */
    convertRouteToTitle(routePath) {
        return routePath.split('-').map((/**
         * @param {?} x
         * @return {?}
         */
        x => x[0].toUpperCase() + x.slice(1))).join(' ');
    }
    /**
     * @protected
     * @param {?} pageRoutes
     * @return {?}
     */
    setProgressSteps(pageRoutes) {
        // Interface for wizard progress items
        this.progressSteps = pageRoutes.map((/**
         * @param {?} page
         * @return {?}
         */
        page => {
            if (page.path !== '') {
                return {
                    title: this.convertRouteToTitle(page.path),
                    route: page.path
                };
            }
        })).filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x));
    }
}
if (false) {
    /**
     * Route items for the stepper
     * @type {?}
     */
    Container.prototype.progressSteps;
}
