// moh-common-lib/esm2015/models/src/mask.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// List of constants used for masking inputs
import { EventEmitter, Input, Output } from '@angular/core';
import { Base } from './base';
/** @type {?} */
export const LETTER = /[A-Z]/i;
// Ignore case here, then upperCase it via pipe.
/** @type {?} */
export const NUMBER = /\d/;
/** @type {?} */
export const SPACE = ' ';
export class MaskModel extends Base {
    constructor() {
        super();
        this.valueChange = new EventEmitter();
        this.disabled = false;
        this.required = true;
    }
    /**
     * Upper cases letters in string
     * @param {?} text
     * @return {?}
     */
    upperCasePipe(text) {
        return text.toUpperCase();
    }
    /**
     * Updates the value
     * @param {?} value
     * @return {?}
     */
    onUpdate(value) {
        // Emit value without spaces
        this.valueChange.emit(value);
    }
}
MaskModel.propDecorators = {
    value: [{ type: Input }],
    valueChange: [{ type: Output }],
    disabled: [{ type: Input }],
    required: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MaskModel.prototype.value;
    /** @type {?} */
    MaskModel.prototype.valueChange;
    /** @type {?} */
    MaskModel.prototype.disabled;
    /** @type {?} */
    MaskModel.prototype.required;
    /** @type {?} */
    MaskModel.prototype.mask;
    /** @type {?} */
    MaskModel.prototype.placeholder;
}
